
```{r only run this chunk once to install packages, include=FALSE, eval=FALSE}

# vegan package requires fortran compiler; see MacBook setup SOP for how to install
install.packages(c("knitr", "here", "tidyverse", "rvest", "pander", "kableExtra", "RColorBrewer", "readxl", "patchwork", "zoo", "ggthemes", "stringdist", "openxlsx", "ggsci", "reticulate", "vegan", "ape", "RSocrata"))

```

```{r manual input ARTIC primer bed file to use}

primer_select <- 2 # set variable as 1 for V3 or 2 for V5.3.2

artic_primer_scheme <- c("V3", "V5.3.2")[primer_select]

# minimum acceptable number of reads per sample
min_reads <- 1000000

project_name <- basename(here())

sequencing_date <- as.Date(gsub("_.*", "", project_name))

```

```{r load libraries}

library(here)
library(tidyverse)
library(rvest)
library(pander)
library(kableExtra)
library(ggsci)
library(ggthemes)
library(RColorBrewer)
library(patchwork)
library(zoo)
library(stringdist)
library(scales)
library(vegan)
library(ape)
library(stringi)

```

```{r load variables, functions, and signature from common files}

tryCatch(
  {
    source(file.path(dirname(here()), "aux_files", "r_scripts", "config", "config_variables.R"))
  },
  error = function(e) {
    stop (simpleError("The config_variables.R file needs to sit in a [aux_files/r_scripts/config] directory path above this project directory"))
  }
)

tryCatch(
  {
    source(file.path(dirname(here()), "aux_files", "r_scripts", "functions", "R_all_functions_v3.R"))
  },
  error = function(e) {
    stop (simpleError("The R_all_functions_v3.R file needs to sit in a [aux_files/r_scripts/functions] directory path above this project directory"))
  }
)

tryCatch(
  {
    signature_fp <- list.files(file.path(dirname(here()), "aux_files", "pdf_generation", "signature"), pattern = ".png|.jpeg|.tif*|.gif", full.names = TRUE)
    tag <- signature_fp[!grepl("blank.png", signature_fp)]
    blank <- signature_fp[grepl("blank.png", signature_fp)]
  },
  error = function(e) {
    stop (simpleError("The signature image file needs to sit in a [aux_files/pdf_generation/signature] directory path above this project directory"))
  }
)

```

```{r file paths for metadata and results}

### mapping file paths
mapping_file_fp <- list.files(here("metadata"), pattern = "_metadata.csv", full.names = TRUE)

### file paths to save results to
unkwn_barcodes_fp <- here("data", paste0(sequencing_date, "_unknown_barcodes.csv"))
fastqc_qual_fp <- here("data", paste0(sequencing_date, "_fastqc_qual.csv"))
actual_read_lengths_fp <- here("data", paste0(sequencing_date, "_read_lengths.csv"))

kraken_props_fp <- here("data", paste0(sequencing_date, "_kraken_props.csv"))
snpdists_fp <- here("data", paste0(sequencing_date, "_snpdists.csv"))
software_version_fp <- here("data", paste0(sequencing_date, "_software_version.csv"))
cecret_results_fp <- here("data", paste0(sequencing_date, "_cecret_results.csv"))
cdc_tracker_fp <- here("data", paste0(sequencing_date, "_cdc_data_tracker_lineages.csv"))
download_cdc_tracker <- !file.exists(cdc_tracker_fp)

```

```{r load metadata}

#define control sample types
ctrl_fct_lvls <- c("Water control", "Reagent control", "Environmental control", "Mock DNA positive control")

s_meta <- read_delim(mapping_file_fp, delim = ",") %>%
  mutate(isControl = sample_group %in% ctrl_fct_lvls) %>%
  mutate(barcode_sequence = paste0(index, "+", index2))

read_length <- unique(s_meta$read_length)
index_length <- unique(s_meta$index_length)

unaccounted_sample_types <- sort(unique(s_meta$ww_group)[!grepl(paste0(c(ctrl_fct_lvls, " - Merged$"), collapse = "|"),
                                                                unique(s_meta$ww_group))])

color_group_lvls <- c(NA, "Reference", "Unassigned reads",
                      ctrl_fct_lvls,
                      unlist(lapply(unaccounted_sample_types, function(x) c(x, paste0(x, " - Poor"), paste0(x, " - Merged")))))

date_col <- sym("sample_collect_date")
sample_collect_date_found <- TRUE

if(!any(grepl("sample_collect_date", colnames(s_meta)))) {
  date_col <- sym("sample_received_date")
  sample_collect_date_found <- FALSE
}

```

<!-- Load QC data -->

```{r find run folder name, eval = import_data}

#get the sequencing run folder
run_folder_name <- list.files(here("data", "processed_bclconvert"), pattern = "^[0-9].*[0-9A-Z]$")

folder_date <- paste0("20", gsub(".*/|_.*", "", run_folder_name)) %>%
  as.Date(format = "%Y%m%d")

if (is.na(folder_date)) {
  stop(simpleError(paste0("Could not find the demultiplexed report from BCL Convert. Have you tried downloading the data?")))
}

if (folder_date != gsub("_.*", "", basename(here()))) {
  stop(simpleError("The run date on the sequencing folder does not match the date of this RStudio project!"))
}

```

```{r file paths for qc data and software versions, eval = import_data}

### demultiplexing results from nfcore
demux_software_fp <- here("data", "processed_bclconvert", "pipeline_info", "software_versions.yml")
demuxsum_fp <- here("data", "processed_bclconvert", run_folder_name, "Reports", "Demultiplex_Stats.csv")
quality_metrics_fp <- here("data", "processed_bclconvert", run_folder_name, "Reports", "Quality_Metrics.csv")
top_unkwn_fp <- here("data", "processed_bclconvert", run_folder_name, "Reports", "Top_Unknown_Barcodes.csv")

cecret_software_fp <- here("data", "processed_cecret", "software_versions.yml")

```

```{r save software versions, eval = import_data}

bcl_version <- read_delim(demux_software_fp, delim = " ", col_names = FALSE) %>%
  filter(grepl("Nextflow: |nf-core/demultiplex: |bclconvert: |falco: ", X1)) %>%
  separate(col = "X1", into = c("software", "version"), sep = ": ") %>%
  mutate(software = gsub("^ *", "", software),
         software = gsub("/", "-", software),
         version = gsub("'", "", version),
         pipeline = "nf-core-demultiplex") %>%
  select(pipeline, software, version)

cecret_workflow_version <- read_delim(cecret_software_fp, delim = " ", col_names = FALSE) %>%
  filter(!grepl("container:", X1),
         grepl("^ ", X1)) %>%
  unique() %>%
  separate(col = "X1", into = c("software", "version"), sep = ": ") %>%
  mutate(software = gsub("^ *", "", software),
         software = gsub("/", "-", software),
         version = gsub("'", "", version),
         pipeline = "UPHL-BioNGS-Cecret",
         software = case_when(software == "workflow" ~ "Cecret",
                              software == "tag" ~ "nextclade-dataset",
                              software == "barcode" ~ "freyja-dataset",
                              TRUE ~ software))

nf_profile_software_version <- read_csv(software_version_fp) %>%
  filter(grepl("kraken2_db|primer_set", software)) %>%
  unique()

nf_artic_version <- nf_profile_software_version %>%
  filter(software == "primer_set") %>%
  select(version) %>%
  mutate(version = gsub("ncov_", "", version)) %>%
  pull()

if(nf_artic_version != artic_primer_scheme) {
    stop(simpleError("The primer_set setting in the Nextflow config file does not match the selected primer set in this RStudio project. Investigate!!"))
}

bcl_version %>%
  rbind(cecret_workflow_version) %>%
  rbind(nf_profile_software_version) %>%
  write_csv(file = software_version_fp)

```

```{r load read numbers and save qc figures, eval = import_data}

quality_metrics <- read_delim(quality_metrics_fp) %>%
  rename(lane = "Lane", sample_id = "SampleID", Q30 = "% Q30") %>%
  select(lane, sample_id, index, index2, ReadNumber, Q30) %>%
  filter(!grepl("^I", ReadNumber)) %>%
  mutate(ReadNumber = paste0("R", ReadNumber, "_Q30")) %>%
  pivot_wider(names_from = "ReadNumber", values_from = "Q30")

merging_columns <- c("barcode_sequence", "index", "index2")

#read counts by sample number
samp_reads <- read_delim(demuxsum_fp) %>%
  rename_all(.funs = tolower) %>%
  rename_all(.funs = function(x) gsub("# ", "", x)) %>%
  rename_all(.funs = function(x) gsub("% ", "percent_", x)) %>%
  rename_all(.funs = function(x) gsub(" ", "_", x)) %>%
  rename(sample_id = "sampleid", barcode_sequence = "index", read_pairs = "reads", percent_of_lane = "percent_reads") %>%
  mutate(sample_id = gsub("_", "-", sample_id),
         barcode_sequence = gsub("-", "+", barcode_sequence),
         read_counts = read_pairs*2) %>%
  merge(quality_metrics, by = c("lane", "sample_id"), all = TRUE) %>%
  merge(select(s_meta, sample_id, uniq_sample_name, merging_columns),
        by = c("sample_id", merging_columns), all = TRUE) %>%
  mutate(across(starts_with("index"), ~replace_na(., "TTTTTTTTTT")),
         across(starts_with("barcode_sequence"), ~replace_na(., "TTTTTTTTTT+TTTTTTTTTT")),
         across(starts_with("percent_of_lane"), ~replace_na(., 0)),
         across(matches("^index|barcode_sequence"), ~ case_when(sample_id == "Undetermined" ~ NA, TRUE ~ (.)))) %>%
  group_by(uniq_sample_name) %>%
  mutate(across(matches("percent_|lane|_Q30"), ~replace_na(., replace = mean(., na.rm = TRUE))),
         across(contains("read"), ~replace_na(., replace = sum(., na.rm = TRUE)))) %>%
  ungroup() %>%
  select(-uniq_sample_name)

top_unkwn_indices <- read_delim(top_unkwn_fp) %>%
  rename_all(.funs = tolower) %>%
  rename_all(.funs = function(x) gsub("# ", "", x)) %>%
  rename_all(.funs = function(x) gsub("% ", "percent_", x)) %>%
  rename_all(.funs = function(x) gsub(" ", "_", x)) %>%
  mutate(barcode_sequence = paste0(index, "+", index2)) %>%
  #get the smallest hamming distance to the sample barcode sequences
  #levenshtein distance is useful for accounting for insertions and deletions between strings, but not needed in this case to determine barcode distances
  mutate(min_dist = sapply(barcode_sequence,
                           function(x) min(stringdist(x, b = samp_reads$barcode_sequence[!is.na(samp_reads$barcode_sequence)], method = "hamming")))) %>%
  rename(read_counts = "reads") %>%
  select(lane, barcode_sequence, read_counts, starts_with("percent"), min_dist) %>%
  order_on_other_col(barcode_sequence, read_counts) %>%
  droplevels()

write_csv(top_unkwn_indices, unkwn_barcodes_fp)

non_ctrl_samples <- s_meta %>%
  filter(!sample_type %in% ctrl_fct_lvls) %>%
  select(sample_id) %>%
  pull()

if(!remove_undetermined_file) {
  #throw error if the top unknown index is greater than 0.5% percent of all reads
  if(max(top_unkwn_indices$percent_of_all_reads) > 0.005 & unique(s_meta$instrument_type) == "MiSeq") {
    stop(simpleError("Top unknown barcode sequence makes up more than 0.5% of all reads. There may be a misassigned barcode"))
  }
  
  #throw error if Undetermined sample has more reads than non-control samples
  if(samp_reads[samp_reads$sample_id == "Undetermined", "read_counts"] >
     sum(samp_reads[samp_reads$sample_id %in% non_ctrl_samples, "read_counts"])) {
    stop(simpleError("Undetermined sample has more reads than the median. Double check the sample barcodes"))
  }
}

```

```{r load fastqc files, eval = import_data}

fastqc_fps <- list.files(here("data", "processed_bclconvert", run_folder_name), pattern = "fastqc_data.txt", recursive = TRUE, full.names = TRUE)
fastqc_fps <- fastqc_fps[!grepl("GenericSampleID", fastqc_fps)]

#line positions to skip to
quality_indic <- ">>Per base sequence quality"
seq_length_indic <- ">>Sequence Length Distribution"

fastqc <- lapply(fastqc_fps, function(f) {
  
  #find the line to start reading the file as tsv
  read_all <- readLines(f)
  line_start <- grep(quality_indic, read_all)
  end_modules <- grep(">>END_MODULE", read_all) - line_start
  line_end <- min(end_modules[which(end_modules > 0)]) - 2
  rm(read_all)
  
  read_qc <- read_tsv(f, skip = line_start, n_max = line_end)
  read_qc$sample_id <- gsub(".*/|_S[0-9]*|_R[12].*", "", f)
  read_qc$direction <- gsub(".*(R[12]).*", "\\1", f)
  return(read_qc)
  }) %>%
  do.call("rbind", .) %>%
  rename(position = "#Base") %>%
  select(sample_id, direction, position, Mean)

length_check <- fastqc %>%
  group_by(sample_id, direction) %>%
  summarize(max_length = max(position)) %>%
  ungroup() %>%
  select(max_length) %>%
  pull()

#throw error if fastqc does not have the max read length
if(!all(length_check == read_length)) {
  stop(simpleError(paste0("Some samples did not have reads up to ", read_length, " bp")))
}

actual_read_lengths <- lapply(fastqc_fps, function(f) {
  #find the line to start reading the file as tsv
  read_all <- readLines(f)
  line_start <- grep(seq_length_indic, read_all)
  end_modules <- grep(">>END_MODULE", read_all) - line_start
  line_end <- min(end_modules[which(end_modules > 0)]) - 2
  rm(read_all)
  
  read_in_lengths <- read_tsv(f, skip = line_start, n_max = line_end)
  read_in_lengths$sample_id <- gsub(".*/|_S[0-9]*|_R[12].*", "", f)
  read_in_lengths$direction <- gsub(".*(R[12]).*", "\\1", f)
  return(read_in_lengths)
  }) %>%
  do.call("rbind", .) %>%
  rename(read_length = "#Length", count = "Count") %>%
  group_by(sample_id, read_length) %>%
  summarize(count = sum(count)) %>%
  ungroup() %>%
  mutate(software = "fastqc") %>%
  select(software, sample_id, read_length, count)

write_csv(fastqc, fastqc_qual_fp)
write_csv(actual_read_lengths, actual_read_lengths_fp)

```

```{r load samtools stats files, eval = import_data}

samtools_stats_fps <- list.files(here("data", "processed_cecret", "samtools_stats"),
                                 pattern = ".stats.txt", recursive = TRUE, full.names = TRUE)

#line positions to skip to
summary_indic <- "# Summary Numbers."

samtools_stat <- lapply(samtools_stats_fps, function(f) {
  
  #find the line to start reading the file as tsv
  read_all <- readLines(f)
  line_start <- grep(summary_indic, read_all)
  line_end <- grep("# First Fragment Qualities.", read_all) - line_start - 2
  rm(read_all)
  
  summary_numbers <- read_tsv(f, skip = line_start, n_max = line_end,
                              col_names = c("SN", "samtools_stat", "samtools_stat_val", "comments"))
  summary_numbers$sample_id <- gsub(".*/|_S[0-9]*|\\.stats\\.txt", "", f)
  return(summary_numbers)
  }) %>%
  do.call("rbind", .) %>%
  mutate(samtools_stat = gsub(" ", "_", samtools_stat),
         samtools_stat = gsub(":", "", samtools_stat),
         samtools_stat = gsub("^", "samtools_", samtools_stat)) %>%
  #select certain rows
  filter(grepl(paste0(c("^samtools_raw_total_sequences$", 
                        "^samtools_filtered_sequences$", 
                        "^samtools_sequences$", 
                        "^samtools_reads_mapped$", 
                        "^samtools_reads_unmapped$", 
                        "^samtools_reads_properly_paired$"), collapse = "|"), samtools_stat)) %>%
  select(sample_id, samtools_stat, samtools_stat_val) %>%
  pivot_wider(names_from = "samtools_stat", values_from = "samtools_stat_val")

read_length_indic <- "# Read lengths\\."

mapped_read_lengths <- lapply(samtools_stats_fps, function(f) {
  
  #find the line to start reading the file as tsv
  read_all <- readLines(f)
  line_start <- grep(read_length_indic, read_all)
  #read file until line_end
  line_end <- grep("# Read lengths - first fragments.", read_all) - line_start - 1
  rm(read_all)
  
  #some samples may have 0 reads
  if(line_end == 0) {
    read_lengths <- data.frame(sample_id = gsub(".*/|_S[0-9]*|\\.stats\\.txt", "", f),
                               RL = "RL",
                               samtools_read_length = 0,
                               count = 0)
  } else {
    read_lengths <- read_tsv(f, skip = line_start, n_max = line_end,
                              col_names = c("RL", "samtools_read_length", "count"))
    read_lengths$sample_id <- gsub(".*/|_S[0-9]*|\\.stats\\.txt", "", f)
  }

  return(read_lengths)
  }) %>%
  do.call("rbind", .) %>%
  mutate(software = "samtools") %>%
  select(software, sample_id, samtools_read_length, count)

samtools_tot_sequences <- mapped_read_lengths %>%
  group_by(sample_id) %>%
  summarize(count = sum(count)) %>%
  ungroup() %>%
  merge(samtools_stat, by = "sample_id", all = TRUE) %>%
  mutate(all_reads_equal = count == samtools_raw_total_sequences)

if(!all(samtools_tot_sequences$all_reads_equal)) {
  stop(simpleError("Error with parsing samtools file. Sum of read length count does not equal total number of reads from samtools"))
}

write_csv(mapped_read_lengths, actual_read_lengths_fp, append = TRUE)

```

```{r load amplicon results, eval = import_data}

amplicon_fps <- list.files(here("data", "processed_cecret", "aci"), pattern = "_amplicon_depth.csv", recursive = TRUE, full.names = TRUE)

amplicon_depth <- amplicon_fps %>%
  data_frame(FileName = .) %>%
  group_by(FileName) %>%
  do(read_csv(.$FileName, col_names = TRUE)) %>%
  ungroup() %>%
  select(-FileName) %>%
  rename(sample_id = "bam") %>%
  mutate(sample_id = gsub("_S[0-9]*|\\.primertrim\\.sorted\\.bam", "", sample_id)) %>%
  rename_with(function(x) paste0("amp_tile_", gsub(".*_", "", gsub("_INSERT$", "", x)), "_artic_", artic_primer_scheme), -sample_id)

```

```{r load coverage results, eval = import_data}

#overall coverage depth files (overlapping regions included in overall coverage)
coverage_fps <- list.files(here("data", "processed_cecret", "samtools_depth"), pattern = ".depth.txt", recursive = TRUE, full.names = TRUE)

slide_window_coverage <- coverage_fps %>%
  data_frame(FileName = .) %>%
  group_by(FileName) %>%
  do(read_delim(.$FileName, col_names = FALSE)) %>%
  ungroup() %>%
  filter(X1 == "MN908947.3") %>%
  rename(position = "X2", map_per_position = "X3") %>%
  select(FileName, map_per_position, position) %>%
  mutate(sample_id = gsub(".*/", "", FileName),
         sample_id = gsub("_S[0-9]*|\\.depth\\.txt", "", sample_id)) %>%
  select(-c(FileName)) %>%
  group_by(sample_id) %>%
  mutate(slide_window_100 = rollmean(map_per_position, k=100, fill = 0, align = 'right')) %>%
  ungroup() %>%
  select(sample_id, position, map_per_position, slide_window_100)

coverage <- slide_window_coverage %>%
  group_by(sample_id) %>%
  summarise(pct_genome_coverage_over_30x = sum(map_per_position >= 30)/n(),
            pct_genome_coverage_over_25x = sum(map_per_position >= 25)/n(),
            pct_genome_coverage_over_20x = sum(map_per_position >= 20)/n(),
            pct_genome_coverage_over_15x = sum(map_per_position >= 15)/n(),
            pct_genome_coverage_over_10x = sum(map_per_position >= 10)/n(),
            pct_genome_coverage_over_5x = sum(map_per_position >= 5)/n(),
            pct_genome_coverage_over_1x = sum(map_per_position >= 1)/n(),
            median_coverage = median(map_per_position),
            mean_coverage = mean(map_per_position)) %>%
  ungroup()

s_gene_coverage <- slide_window_coverage %>%
  filter(position >= 21563) %>%
  filter(position <= 25384) %>%
  group_by(sample_id) %>%
  summarise(pct_s_gene_coverage_over_30x = sum(map_per_position >= 30)/n(),
            pct_s_gene_coverage_over_25x = sum(map_per_position >= 25)/n(),
            pct_s_gene_coverage_over_20x = sum(map_per_position >= 20)/n(),
            pct_s_gene_coverage_over_15x = sum(map_per_position >= 15)/n(),
            pct_s_gene_coverage_over_10x = sum(map_per_position >= 10)/n(),
            pct_s_gene_coverage_over_5x = sum(map_per_position >= 5)/n(),
            pct_s_gene_coverage_over_1x = sum(map_per_position >= 1)/n(),
            median_s_gene_coverage = median(map_per_position),
            mean_s_gene_coverage = mean(map_per_position)) %>%
  ungroup()

```

```{r load samtools coverage data, eval = import_data}

#number of reads that mapped to reference
cov_num_fps <- list.files(here("data", "processed_cecret", "samtools_coverage"), pattern = ".cov.txt", recursive = TRUE, full.names = TRUE)

#the mean coverage from samtools is slightly different than by manually calculation but the 1X coverage is the same
cov_stats <- cov_num_fps %>%
  data_frame(FileName = .) %>%
  group_by(FileName) %>%
  do(read_delim(.$FileName, col_names = TRUE)) %>%
  ungroup() %>%
  rename(sample_id = "sample", aligned_reads = "numreads") %>%
  mutate(sample_id = gsub("_S[0-9]*$", "", sample_id)) %>%
  select(sample_id, aligned_reads)

```

<!-- Load analysis data -->

```{r load kmer results, eval = import_data}

k2_files_fps <- list.files(here("data", "processed_cecret", "kraken2"), pattern = "_kraken2_report.txt", recursive = TRUE, full.names = TRUE)

k2_raw <- k2_files_fps %>%
  data_frame(FileName = .) %>%
  group_by(FileName) %>%
  do(read_delim(.$FileName,
                col_names = c("k2_percent", "k2_clade_kmers", "k2_kmers", "k2_rank", "k2_ncbi_taxon_id", "k2_name"),
                col_types = cols("k2_percent" = col_character(),
                                 "k2_clade_kmers" = col_double(),
                                 "k2_kmers" = col_double(),
                                 "k2_ncbi_taxon_id" = col_double()))) %>%
  ungroup() %>%
  mutate(sample_id = gsub(".*/", "", FileName),
         sample_id = gsub("_kraken2_report.txt", "", sample_id),
         k2_name = gsub("^ *", "", k2_name),
         k2_rank = tolower(k2_rank)) %>%
  select(sample_id, k2_rank, k2_name, k2_kmers) %>%
  group_by(sample_id) %>%
  mutate(k2_top_rank = case_when(grepl("^u$|^r$|^d$|^k$|^p$", k2_rank) ~ paste0(k2_rank, "__", k2_name),
                                 TRUE ~ NA)) %>%
  fill(k2_top_rank, .direction = "down") %>%
  #filter for taxon with at least 2 kmer hits
  filter(k2_kmers > 1) %>%
  mutate(k2_props = k2_kmers/sum(k2_kmers)) %>%
  ungroup() %>%
  mutate(k2_taxon = ifelse(k2_top_rank == paste0(k2_rank, "__", k2_name),
                           k2_top_rank,
                           paste0(k2_top_rank, " ", gsub("*[0-9]*", "", k2_rank), "__", k2_name))) %>%
  select(sample_id, k2_taxon, k2_props, k2_kmers) %>%
  mutate(sample_id = gsub("_S[0-9]*$", "", sample_id))

k2_sc2 <- k2_raw %>%
  # includes SARSr-CoV species and SARS-CoV-2 strain
  filter(grepl("Severe acute respiratory syndrome", k2_taxon)) %>%
  group_by(sample_id) %>%
  summarize(k2_sc2_kmers = sum(k2_kmers),
            k2_sc2_props = sum(k2_props)) %>%
  ungroup()

k2_raw %>%
  write_csv(kraken_props_fp)

```

```{r load NextClade results, eval = import_data}

nc_fp <- here("data", "processed_cecret", "nextclade", "nextclade.tsv")

next_clade <- nc_fp %>%
  read_delim(col_names = TRUE,
             col_types = cols(missing = col_character(),
                              clade_nextstrain = col_character(),
                              clade = col_character())) %>%
  rename(sample_id = "seqName",
         nextclade_clade = "clade",
         nextclade_clade_who = "clade_who",
         nextclade_qc_overallstatus = "qc.overallStatus",
         nextclade_qc_overallscore = "qc.overallScore",
         nextclade_index = "index") %>%
  mutate(sample_id = gsub("_S[0-9]*$", "", sample_id))

```

```{r load freyja results, eval = import_data}

freyja_fps <- list.files(here("data", "processed_cecret", "freyja"), pattern = "_demix.tsv", recursive = TRUE, full.names = TRUE)

freyja <- freyja_fps %>%
  data_frame(FileName = .) %>%
  group_by(FileName) %>%
  do(read_tsv(.$FileName,
              col_names = c("col_names", "values"),
              col_types = cols("col_names" = col_character(),
                               "values" = col_character()))) %>%
  ungroup() %>%
  filter(!is.na(col_names)) %>%
  mutate(sample_id = gsub(".*/", "", FileName),
         sample_id = gsub("_S[0-9]*|_demix.tsv", "", sample_id),
         col_names = paste0("freyja_", col_names)) %>%
  select(sample_id, col_names, values) %>%
  pivot_wider(names_from = "col_names", values_from = "values")
  
```

```{r load snp dists results, eval = import_data}

snpdists_raw_fp <- here("data", "processed_cecret", "snp-dists", "snp-dists.txt")

snpdists <- snpdists_raw_fp %>%
  read_csv() %>%
  rename(sample_id = 1) %>%
  mutate(sample_id = gsub("_S[0-9]*$", "", sample_id)) %>%
  rename_all(.funs = function(x) gsub("_S[0-9]*$", "", x)) %>%
  pivot_longer(cols = -sample_id, names_to = "compared_to", values_to = "snps")

snpdist_check <- snpdists %>%
  filter(sample_id == compared_to)

if(!all(snpdist_check$snps == 0)) {
  stop(simpleError("Something wrong with the snp-dist process"))
}
  
snpdists %>%
  write_csv(snpdists_fp)
  
```

```{r load cecret results, eval = import_data}

# remove these software versions from the cecret results
cecret_software_select <- c("ivar", "ivar consensus", "bwa", "seqyclean", "Cecret version", "pangolin_scorpio_version", "pangolin_constellation_version", "pangolin_pangolin_version", "pangolin_version")

#pangolin data is included with this summary file
cecret_aggregate <- here("data", "processed_cecret", "cecret_results.csv") %>%
  read_csv() %>%
  mutate(sample_number = as.numeric(gsub("^.*_S", "", sample_id))) %>%
  select(-c(sample_id, samtools_meandepth_after_trimming, samtools_per_1X_coverage_after_trimming, fasta_line)) %>%
  rename(sample_id = "sample",
         vadr_passfail = "vadr_p/f",
         nextclade_qc_overallscore = "nextclade_qc.overallScore",
         nextclade_qc_overallstatus = "nextclade_qc.overallStatus") %>%
  as.data.frame() %>%
  select(-all_of(cecret_software_select))

```

```{r write cecret results, eval = import_data}

list(cov_stats, coverage, s_gene_coverage, next_clade, freyja, amplicon_depth, samtools_stat, k2_sc2) %>%
  reduce(merge, by = "sample_id", all = TRUE) %>%
  merge(cecret_aggregate,
        by = c("sample_id", "nextclade_clade", "nextclade_clade_who", "nextclade_qc_overallstatus", "nextclade_qc_overallscore"),
        all = TRUE,
        sort = FALSE) %>%
  mutate(nextclade_clade_who = case_when((grepl("^B$|^B\\.[0-9]*$", Nextclade_pango) & nextclade_clade == "19A") ~ "Reference", #Illumina PC
                                         (grepl("^A$|^A\\.[0-9]*$", Nextclade_pango) & nextclade_clade == "19B") ~ "Reference", #ZeptoMetrix PC
                                         TRUE ~ nextclade_clade_who),
         sample_id = gsub("_S[0-9]*$", "", sample_id),
         coverage_qc_status = case_when(pct_genome_coverage_over_10x >= 0.6 ~ "Good",
                                        pct_genome_coverage_over_10x < 0.6 ~ "Poor",
                                        TRUE ~ NA)) %>%
  merge(samp_reads, by = "sample_id", all = TRUE) %>%
  write_csv(cecret_results_fp)
  
```

```{r read in data}

top_unkwn_indices <- read_csv(unkwn_barcodes_fp)

fastqc <- read_csv(fastqc_qual_fp)

actual_read_lengths <- read_csv(actual_read_lengths_fp)

cecret_results <- read_csv(cecret_results_fp)

software_version <- read_csv(software_version_fp) %>%
  select(software, version) %>%
  deframe() %>%
  as.list()

k2 <- read_csv(kraken_props_fp)

snpdists <- read_csv(snpdists_fp)

```

```{r merge metadata with data and select samples with results}

#df with all samples
s <- s_meta %>%
  merge(cecret_results, by = c("sample_id", "barcode_sequence", "index", "index2"), all = TRUE) %>%
  mutate(ww_group = case_when(grepl("^Undetermined$", sample_id) ~ "Unassigned reads",
                              ((pct_genome_coverage_over_10x < 0.6 | is.na(pct_genome_coverage_over_10x)) &
                                 !(isControl | grepl("- Merged$", ww_group))) ~ paste0(ww_group, " - Poor"),
                              TRUE ~ ww_group),
         ww_group = factor(ww_group, levels = color_group_lvls),
         sample_group = case_when(grepl("^Undetermined$", sample_id) ~ "Unassigned reads",
                                  TRUE ~ sample_group),
         sample_group = factor(sample_group, levels = unique(sample_group[order(ww_group)])),
         date_facet = ifelse(grepl("Unassigned reads|control", ww_group), "Control", as.character(!!date_col)),
         date_facet = factor(date_facet, levels = c("Control", sort(unique(date_facet[!grepl("Control", date_facet)])))),
         #order the sample_id by sample type then by plate position
         sample_id = factor(sample_id, levels = unique(sample_id[str_order(paste0(as.numeric(sample_group), "-", grepl("Merged", sample_id), "-", !!date_col), decreasing = FALSE, numeric = TRUE)]))) %>%
  arrange(sample_id) %>%
  droplevels() %>%
  filter(!is.na(sample_group))

has_N2 <- any(grepl("^N2_gc/uL$", colnames(s)))

has_qubit <- !all(is.na(s$qubit_conc_ng_ul))

if(any(is.na(s$ww_group))) {
  stop(simpleError("A sample type is unaccounted for either in the metadata sheet or in color_group_lvls"))
}

```

<!-- Prepare files containing samples with good results for uploading -->

```{r download pango names and CDC Tracker lineages, eval = download_cdc_tracker & report_only_cdc_lineages}

sixty_days_ago <- Sys.Date() - 60
cdc_tracker_url <- paste0("https://data.cdc.gov/resource/jr58-6ysp.json?$select=variant,week_ending&$where=week_ending>'", sixty_days_ago, "'")

dl_cdc_names <- simpleError("Initial message")

counter <- 1
max_tries <- 5

while(inherits(dl_cdc_names, "error") & counter <= max_tries) {
  message("Trying SoQL query from https://data.cdc.gov/\nAttempt #", counter)
  Sys.sleep(counter)
  dl_cdc_names <- tryCatch(
    {
      RSocrata::read.socrata(cdc_tracker_url)
    }, 
    error = function(error_input) {
      error_input
    }
  )
  counter <- counter + 1
}

if(counter > max_tries) {
  stop(simpleError(paste0("\nThe lineages from CDC COVID Data Tracker could not be downloaded",
                          "\nDownload the CDC json file using the following URL:",
                          "\nhttps://data.cdc.gov/resource/jr58-6ysp.json?$query=SELECT%0A%20%20%60variant%60%2C%0A%20%20count(%60variant%60)%20AS%20%60count_variant%60%2C%0A%20%20min(%60creation_date%60)%20AS%20%60min_creation_date%60%2C%0A%20%20max(%60creation_date%60)%20AS%20%60max_creation_date%60%0AGROUP%20BY%20%60variant%60",
                          "\nSave the file as:\n", cdc_tracker_fp,
                          "\nUncomment the code below to read in this CDC variant list")))
}

# dl_cdc_names <- read_delim(cdc_tracker_fp, delim = "\",\"", col_names = c("variant", "count_variant", "min_creation_date", "max_creation_date")) %>%
#   mutate(across(everything(), ~ gsub(".*\":\"", "", .)),
#          week_ending = Sys.Date()) %>%
#   select(variant, week_ending)

cdc_lineages <- dl_cdc_names %>%
  unique() %>%
  mutate(week_ending = as.Date(week_ending),
         week_rank = factor(week_ending, levels = sort(unique(week_ending), decreasing = TRUE))) %>%
  #filter for the two most recent weeks
  filter(week_rank %in% levels(week_rank)[1:2]) %>%
  arrange(week_ending, variant) %>%
  select(-week_rank)

cdc_lineages %>%
  write_csv(cdc_tracker_fp)

```

```{r prepare freyja table}

flodel <- function(word1, word2) {
   # the length of the 2nd word
   n <- nchar(word2)
   # two vectors of characters of the same length n
   c1 <- strsplit(word1, "", fixed = TRUE)[[1]][1:n]
   c2 <- strsplit(word2, "", fixed = TRUE)[[1]][1:n]
   # a vector that is TRUE as long as the characters match
   m <- as.logical(cumprod(c1 == c2) & !is.na(cumprod(c1 == c2)))
   # the answer
   a <- paste(c1[m], collapse = "")
   # return answer if ends with period and matches length of word2
   if(sum(m) == n) {
     a
   }
   else {
     ""
   }
}

freyja_lineages <- s %>%
  select(sample_id, sample_name, uniq_sample_name, sample_group, ww_group, !!date_col, sequencing_date,
         isControl, freyja_lineages, freyja_abundances, pct_genome_coverage_over_10x) %>%
  group_by(sample_id) %>%
  mutate(lineages = list(freyja_lineages),
         relative_abundances = list(freyja_abundances)) %>%
  separate_rows(c(lineages, relative_abundances), sep=" ") %>%
  mutate(lineages = gsub("\\([0-9]*\\)|-like", "", lineages),
         lineages = ifelse((is.na(lineages) | lineages == "Misc" | relative_abundances == "nan" | relative_abundances == "inf"),
                           "Unresolved", lineages),
         relative_abundances = as.numeric(gsub("nan|inf", "", relative_abundances)),
         main_lineage = gsub("\\..*", "", lineages)) %>%
  ungroup() %>%
  select(-c("freyja_lineages", "freyja_abundances")) %>%
  unique() %>%
  arrange(sample_id, relative_abundances, str_rank(lineages, numeric = TRUE))

if(report_only_cdc_lineages) {
  
  lineages_to_keep <- read_csv(cdc_tracker_fp) %>%
    filter(!is.na(variant)) %>%
    select(variant) %>%
    unique()
  
  # this dataframe groups the lineages into CDC lineages, otherwise it will group as Other
  freyja_grouping <- freyja_lineages %>%
    mutate(lineages = gsub("$", ".", lineages),
           longest_common_substring = sapply(lineages,
                                             function(x) max(sapply(gsub("$", ".", c("A", "B", lineages_to_keep$variant)),
                                                                    function(y) flodel(x, y)))),
           relative_abundances = ifelse(is.na(relative_abundances) & grepl("Unresolved", lineages), 1, relative_abundances),
           lineages = gsub("\\.$", "", lineages),
           CDC_lineages = fct_other(gsub("\\.$", "", longest_common_substring),
                                    keep = c("A", "B", lineages_to_keep$variant),
                                    other_level = "Other lineages"),
           CDC_lineages = factor(CDC_lineages, levels = c(str_sort(lineages_to_keep$variant, numeric = TRUE), "Other lineages", "A", "B"))) %>%
    droplevels()
  
} else{
  
  # this dataframe groups the sublineages with the same parent lineages together and gets the sum of the abundances (if the abundances are all equal)
  freyja_grouping <- freyja_lineages %>%
    group_by(sample_id, main_lineage, relative_abundances) %>%
    #add an extra period to the end to help with matching lineages correctly
    mutate(lineages = gsub("$", ".", lineages),
           least_common_substring = attr(adist(lineages, counts = TRUE), "trafos")[, 1],
           end_match = do.call(rbind, stringi::stri_locate_all_regex(least_common_substring, "^M+"))[, 2],
           has_lowest_str_length = str_length(lineages) == min(str_length(lineages)),
           substring_end = ifelse(sum(has_lowest_str_length) == 1 &
                                  grepl("^M+$", least_common_substring), end_match, NA),
           substring_end = ifelse(all(is.na(substring_end)), min(end_match), substring_end)) %>%
    fill(substring_end, .direction = "down") %>%
    ungroup() %>%
    mutate(CDC_lineages = stringi::stri_sub(lineages, 1, substring_end),
           CDC_lineages = gsub("\\.$", "", CDC_lineages),
           CDC_lineages = factor(CDC_lineages, levels = str_sort(unique(CDC_lineages), numeric = TRUE)),
           relative_abundances = ifelse(is.na(relative_abundances) & grepl("Unresolved", CDC_lineages),
                                        1, relative_abundances),
           lineages = gsub("\\.$", "", lineages))
  
  lineages_to_keep <- freyja_grouping %>%
    select(variant = "CDC_lineages") %>%
    filter(variant != "Unresolved") %>%
    unique()
}

freyja_matrix <- freyja_grouping %>%
  complete(nesting(sample_id, ww_group), lineages, fill = list(relative_abundances = 0)) %>%
  group_by(sample_id, lineages) %>%
  summarize(relative_abundances = sum(relative_abundances)) %>%
  ungroup() %>%
  pivot_wider(names_from = "lineages", values_from = "relative_abundances") %>%
  column_to_rownames("sample_id")

# this table groups the individual samples together and calculates the average
freyja_average <- freyja_grouping %>%
  filter(!isControl) %>%
  droplevels() %>%
  mutate(isMerged = ifelse(grepl("Merged$", ww_group), "Merged sample", "Individual samples")) %>%
  complete(nesting(sample_id, !!date_col, pct_genome_coverage_over_10x, sample_group, isMerged),
           nesting(lineages, main_lineage, CDC_lineages),
           fill = list(relative_abundances = 0)) %>%
  group_by(sample_id, !!date_col, pct_genome_coverage_over_10x, sample_group,
           isMerged, lineages, CDC_lineages, main_lineage) %>%
  summarize(relative_abundances = sum(relative_abundances)) %>%
  ungroup() %>%
  filter(!is.na(relative_abundances)) %>%
  group_by(sample_group, !!date_col, isMerged) %>%
  mutate(total_samples = n_distinct(sample_id),
         mean_cov = mean(pct_genome_coverage_over_10x)) %>%
  ungroup() %>%
  # to calculate the total standard deviation of a combined distribution,
  # get the total variance and square root it
  group_by(sample_group, !!date_col, mean_cov, isMerged, lineages, CDC_lineages, main_lineage) %>%
  mutate(avg_rel_abund = relative_abundances/total_samples,
         variance = var(avg_rel_abund)) %>%
  summarize(avg_rel_abund = sum(avg_rel_abund),
            total_sd = sqrt(sum(variance))) %>%
  ungroup() %>%
  droplevels()

```

```{r export data for epi}

#write freyja results for epi's
dir.create(here("upload", "epi"), recursive = TRUE)
epi_report_fp <- here("upload", "epi", paste0(project_name, "_PHL2_WW_sequencing_results.csv"))

#filter s_toReport to remove samples that have A lineage present
samples_with_A <- freyja_grouping %>%
  filter(CDC_lineages == "A") %>%
  select(sample_id) %>%
  unique() %>%
  pull()

#only get merged samples that have passing results and controls
s_toReport <- s %>%
  filter(!grepl("None|Undetermined", sample_id),
         coverage_qc_status == "Good",
         grepl("Wastewater sample - Merged", ww_group),
         !sample_id %in% samples_with_A) %>%
  mutate(sample_id = as.character(sample_id)) %>%
  droplevels()

# export results to epi only if sample_collect_date is present
if(sample_collect_date_found) {
  
  freyja_summary <- freyja_grouping %>%
    group_by(sample_id, uniq_sample_name, sample_group, isControl,
             sequencing_date, sample_collect_date,
             pct_genome_coverage_over_10x, CDC_lineages) %>%
    summarize(relative_abundances = sum(relative_abundances)) %>%
    ungroup() %>%
    filter(relative_abundances >= 0.01) %>%
    arrange(sample_collect_date, sample_group, desc(relative_abundances), CDC_lineages)

  freyja_summary %>%
    filter(sample_id %in% s_toReport$sample_id,
           !isControl,
           CDC_lineages %in% lineages_to_keep$variant) %>%
    select(sample_collect_date, sample_group, freyja_lineage = "CDC_lineages") %>%
    write_csv(epi_report_fp)
  
} else {
  data.frame(sample_collect_date = "",
             sample_group = "",
             freyja_lineage = "") %>%
    write_csv(epi_report_fp)
}

```

```{r file for lims}

#write file for LIMS
dir.create(here("upload", "lims"), recursive = TRUE)
lims_import_fp <- here("upload", "lims", paste0(project_name, "_PHL2_WW_LIMS_import.csv"))

fields4lims <- c("sample_id",
                 "sample_number",
                 "sample_name",
                 "uniq_sample_name",
                 "sample_group",
                 "ww_group", 
                 "sample_type",
                 "environmental_site",
                 "sample_collected_by",
                 "sample_collect_date",
                 "sample_received_date",
                 "ddpcr_analysis_date",
                 "plate",
                 "plate_row",
                 "plate_col",
                 "plate_coord",
                 "idt_set",
                 "idt_plate_row",
                 "idt_plate_col",
                 "idt_plate_coord",
                 "UDI_Index_ID",
                 "index",
                 "index2",
                 "prj_descrip",
                 "sequencing_date",
                 "instrument_type",
                 "read_length",
                 "index_length",
                 "run_cd",
                 "run_q30",
                 "run_pf",
                 "run_error",
                 
                 #results and qc
                 "read_counts",
                 "R1_Q30",
                 "R2_Q30",
                 "aligned_reads",
                 
                 "median_coverage",
                 "mean_coverage",
                 "k2_sc2_kmers",
                 "k2_sc2_props",
                 "pct_genome_coverage_over_1x",
                 "pct_genome_coverage_over_10x",
                 "freyja_cdc_lineages",
                 "freyja_cdc_lineages_abund", 
                 "A_detected")

select_software <- read_csv(software_version_fp) %>%
  filter(grepl(paste("^Nextflow",
                     "nf-core-demultiplex",
                     "bclconvert",
                     "Cecret",
                     "primer_set",
                     "kraken2",
                     "kraken2_db",
                     "bwa",
                     "samtools",
                     "ivar", 
                     "freyja",
                     "freyja-dataset$", sep = "$|^"), software)) %>%
  mutate(software_version = paste0(pipeline, "_", software)) %>%
  select(software_version, version) %>%
  pivot_wider(names_from = "software_version", values_from = "version")

freyja_summarized <- freyja_summary %>%
  mutate(CDC_lineages = factor(CDC_lineages, levels = c(tail(levels(freyja_summary$CDC_lineages), 2),
                                                        head(levels(freyja_summary$CDC_lineages), -2))),
         relative_abundances = round(relative_abundances, 3)) %>%
  arrange(CDC_lineages, desc(relative_abundances)) %>%
  group_by(sample_id, uniq_sample_name, sequencing_date, pct_genome_coverage_over_10x) %>%
  summarize(freyja_cdc_lineages = paste0(CDC_lineages, collapse = "; "),
            freyja_cdc_lineages_abund = paste0(CDC_lineages, "-", relative_abundances, collapse = "; ")) %>%
  ungroup()

s_merged <- freyja_summarized %>%
  filter(grepl("Merged", sample_id)) %>%
  # sometimes the merged Zepto will not have A lineage, even when one of its replicates have it
  mutate(merged_A_detected = sample_id %in% samples_with_A) %>%
  select(sequencing_date,
         uniq_sample_name,
         merged_pct_genome_coverage_over_10x = "pct_genome_coverage_over_10x",
         merged_freyja_cdc_lineages = "freyja_cdc_lineages",
         merged_freyja_cdc_lineages_abund = "freyja_cdc_lineages_abund",
         merged_A_detected)

lims_import <- s %>%
  filter(!grepl("Merged|Undetermined", sample_id)) %>%
  left_join(freyja_summarized, by = c("sample_id", "uniq_sample_name", "sequencing_date", "pct_genome_coverage_over_10x")) %>%
  left_join(s_merged, by = c("uniq_sample_name", "sequencing_date")) %>%
  mutate(A_detected = sample_id %in% samples_with_A) %>%
  select(all_of(fields4lims), starts_with("merged_")) %>%
  cbind(select_software) %>%
  arrange(sample_number)

if(sample_collect_date_found) {
  lims_import %>%
    write_csv(lims_import_fp)
}

```

```{r export data for reporting and submission}

#see dx.doi.org/10.17504/protocols.io.ewov14w27vr2/v10 for the ww submission guidance
# Biosamples are created at the extraction level (each RNA extraction will get a unique BioSample ID). Create a base ID (isolate) for each sample collection (for example, LABID_001), then add an index (specimen_processing_id) to represent each extraction (e.g. LABID_001.01). Every Sample Name from a single Submitter must be unique. 

#read in data submission metadata
data_submit_metadata_fp <- file.path(dirname(here()), "aux_files", "data_submission", "ww", "wastewater_data_submission_metadata.csv")

data_submit_metadata <- generate_common_metadata_fields(data_submit_metadata_fp) %>%
  select(!starts_with(c("nwss-", "gb-"))) #currently no recommendation to submit wastewater data to GenBank
#see why there warning here

dir.create(here("upload", "seqsender"), recursive = TRUE)

fasta_seqsender_meta_fp <- here("upload", "seqsender", paste(project_name, "PHL2_seqsender_fasta_upload_orig.csv", sep = "_"))
fastq_seqsender_meta_fp <- here("upload", "seqsender", paste(project_name, "PHL2_seqsender_fastq_upload_orig.csv", sep = "_"))

seqsender <- s %>%
  filter(uniq_sample_name %in% s_toReport$uniq_sample_name,
         !isControl)

if(nrow(seqsender) != 0 & sample_collect_date_found) {

  seqsender <- seqsender %>%
    rename(sequence_name = "sample_id",
           `gs-covv_coverage` = "median_coverage",
           `gs-covv_orig_lab` = "sample_collected_by",
           `bs2-isolation_source` = "sample_type",
           `bs17-ww_surv_system_sample_id` = "bs-ww_surv_system_sample_id") %>%
    mutate(collection_date = case_when(!(is.na(sample_collect_date) |
                                         as.character(sample_collect_date) == "Unknown") ~ as.character(sample_collect_date),
                                       !is.na(sample_received_date) ~ as.character(sample_received_date),
                                       TRUE ~ as.character(sequencing_date)),
           collection_date = as.Date(collection_date, tryFormats = c("%Y-%m-%d", "%m/%d/%y", "%m/%d/%Y")),
           collection_year = format(collection_date, format = "%Y"),
           #GISAID fields
           #GISAID name format for WW samples needs to be hCoV-19/env/[country]/[state]-[sample_id without a year]/sample collection year (collection year may be different from sequence year)
           `gs-sample_name` = paste0("hCoV-19/env/USA/PA-", gsub(str_sub(project_name, 1, 4), "", sequence_name), "/", collection_year),
           `gs-covv_seq_technology` = case_when(instrument_type == "MiSeq" ~ "Illumina MiSeq",
                                                instrument_type == "NextSeq2000" ~ "Illumina NextSeq 2000",
                                                TRUE ~ NA_character_),
           `gs-covv_assembly_method` = paste0("UPHL-BioNGS Cecret ", software_version$Cecret),
           `gs-covv_subm_sample_id` = `bs17-ww_surv_system_sample_id`,
           #the following columns are in the submission metadata sheet
           #gs-covv_type
           #gs-covv_passage
           #gs-covv_location
           #gs-covv_add_location
           #gs-covv_host
           #gs-covv_gender
           #gs-covv_patient_age
           #gs-covv_patient_status
           #gs-covv_specimen
           #gs-covv_orig_lab_addr
           #gs-covv_subm_lab
           #gs-covv_subm_lab_addr
           #gs-covv_comment
           #gs-comment_type
           
           #potential GenBank fields
           #`gb-sample_name` = sequence_name,
           #`gb-src-isolate` = paste0("SARS-CoV-2/Wastewater/USA/", sequence_name, "/", collection_year),
           
           #BioSample fields
           `bs1-sample_name` = paste0("SARS-CoV-2/Wastewater/USA/", sequence_name, "/", collection_year),
           #bs2-isolation_source #this column is defined above, not in metadata sheet
           #bs3-geo_loc_name
           `bs4-collected_by` = `gs-covv_orig_lab`,
           #bs5-sequenced_by
           #bs6-purpose_of_ww_sampling
           #bs7-purpose_of_ww_sequencing
           `bs8-collection_site_id` = wwtp_name,
           `bs9-isolate` = uniq_sample_name,
           `bs10-specimen_processing_id` = sample_name,
           #bs11-specimen_processing
           #bs12-specimen_processing_details
           #bs13-concentration_method
           #bs14-extraction_method
           #bs15-extraction_control
           #bs16-project_name
           #bs17-ww_surv_system_sample_id #this column is defined above, not in metadata sheet
           #bs18-ww_sample_site
           #bs19-ww_sample_matrix
           #bs20-ww_sample_type
           #bs21-ww_sample_duration
           #bs22-ww_surv_jurisdiction
           `bs23-ww_population` = population_served,
           #bs24-ww_surv_target_1
           #bs25-ww_surv_target_1_gene
           `bs26-ww_surv_target_1_known_present` = ifelse(has_N2,
                                                          ifelse(N2_present & !is.na(N2_present), "Yes", "No"),
                                                          ""),
           `bs27-ww_surv_target_1_conc` = ifelse(is.na(`N2_gc/L`), "", round(`N2_gc/L`, 2)),
           #bs28-ww_surv_target_1_conc_unit
           #bs29-ww_surv_target_1_protocol
           #bs30-ww_surv_target_2
           #bs31-ww_surv_target_2_gene
           `bs32-ww_surv_target_2_known_present` = ifelse(has_N2,
                                                          ifelse(E_present & !is.na(E_present), "Yes", "No"),
                                                          ""),
           `bs33-ww_surv_target_2_conc` = ifelse(is.na(`E_gc/L`), "", round(`E_gc/L`, 2)),
           #bs34-ww_surv_target_2_conc_unit
           #bs35-ww_surv_target_2_protocol
           `bs36-gisaid_virus_name` = `gs-sample_name`,
           #bs37-gisaid_accession
           #bs38-sample_title
          
           #SRA fields
           `sra1-sample_name` = paste0("SARS-CoV-2/Wastewater/USA/", sequence_name),
           `sra2-library_name` = paste0("PHL2-WW-COVIDSeq-", sequencing_date),
           `sra3-instrument_model` = gsub("^Illumina NextSeq", "NextSeq", `gs-covv_seq_technology`),
           #sra4-library_strategy
           #sra5-library_source
           #sra6-library_selection
           #sra7-library_layout
           #sra8-platform
           #sra9-enrichment_kit
           `sra10-amplicon_PCR_primer_scheme` = paste0("SARS-CoV-2 ARTIC ", artic_primer_scheme),
           #sra11-library_preparation_kit
           `sra12-design_description` = paste0("Genomic reads were enriched using a tiled amplicon strategy with the ARTIC Network Protocol ",
                                               artic_primer_scheme, " primer scheme. Sequencing was performed on the ", `gs-covv_seq_technology`,
                                               " with 2x", read_length, " bp chemistry. Library was prepared using the Illumina COVIDSeq Assay."),
           #sraX-sample_title #include when this works in the seqsender package
           #sra13-file_location
           #sra14-filetype
           `sra15-file_1` = paste0(sequence_name, "_S", sample_number, "_filtered_R1.fastq.gz"),
           `sra16-file_2` = paste0(sequence_name, "_S", sample_number, "_filtered_R2.fastq.gz")) %>%
    arrange(sample_number) %>%
    cbind(data_submit_metadata) %>%
    select(sequence_name, organism, collection_date, ww_group,
           matches("^gs-", ignore.case = FALSE),
           #order the column names by the numbers 1 to 100
           starts_with(paste0("bs", 1:1000, "-"), ignore.case = FALSE),
           starts_with(paste0("sra", 1:1000, "-"), ignore.case = FALSE)) %>%
    rename_at(vars(starts_with("bs")), ~ gsub("bs[0-9]+-", "bs-", .)) %>%
    rename_at(vars(starts_with("sra")), ~ gsub("sra[0-9]+-", "sra-", .))
  
  seqsender %>%
    filter(ww_group == "Wastewater sample - Merged") %>%
    select(-ww_group) %>%
    write_csv(fasta_seqsender_meta_fp, na = "")

  seqsender %>%
    filter(ww_group != "Wastewater sample - Merged") %>%
    select(-ww_group) %>%
    write_csv(fastq_seqsender_meta_fp, na = "")
  
}

#TODO: incorporate the flow data; see email

```

```{r reported numbers}

#double checking to see if number of samples match
n_samples_report <- s_toReport %>%
  filter(!isControl) %>%
  nrow()

#these are samples that failed in the pipeline for whatever reason. Can rerun pipeline to generate results for these samples
rerun_samples <- s %>%
  filter(is.na(pangolin_qc_status)) %>%
  filter(!isControl) %>%
  select(sample_id) %>%
  pull()

#number of epi samples
epi_2_upload_fp <- list.files(here("upload", "epi"), pattern = ".csv", recursive = TRUE, full.names = TRUE)

reported_samples <- epi_2_upload_fp %>%
  read_csv() %>%
  select(sample_name = "sample_group", sample_collect_date) %>%
  filter(!is.na(sample_name)) %>%
  unique() %>%
  nrow()

#number of samples to submit
#check the dcipher file to upload
dcipher_2_upload_fp <- list.files(here("upload", "seqsender"), pattern = "_fasta_upload_orig.csv", recursive = TRUE, full.names = TRUE)

if(length(dcipher_2_upload_fp) != 0) {
  all_reported_samples <- dcipher_2_upload_fp %>%
    read_csv() %>%
    nrow()
} else {
  all_reported_samples <- 0
}

```

<!-- Assign colors to sample attributes and create some ggplot legends and figures -->

```{r sample type attributes for ggplot}

sample_type_num_vector <- 1:length(color_group_lvls)

ann_geom_values <- list()

# pick sample type colors for ggplot
sample_type_color_select <- c(15, 19, 16, 18, 2, #colors for sequencing controls
                              11, 14, 13, #colors for wastewater controls
                              1, 4, 3, #colors for wastewater samples
                              10, 7, 20, 17) #some additional colors

sample_type_color_scheme <- c("#808080", "#000000",
                              pal_d3("category20")(20)[c(sample_type_color_select,
                                                         which(!1:20 %in% sample_type_color_select))])

ann_geom_values[["sample_type_colors"]] <- setNames(sample_type_color_scheme[sample_type_num_vector], color_group_lvls)

# pick shapes for ggplot
sample_group_index <- which(levels(s$sample_group) == "ZeptoSC2"):length(levels(s$sample_group))
ann_geom_values[["sample_group_shapes"]] <- setNames(c(23, 25, 22, 21, 24),
                                                     levels(s$sample_group)[sample_group_index])

# pick alpha values for ggplot
alpha_select <- c(1, 1, 1, 0.4, 1, 0.4, 1, 0.2, 0.2)
additional_alphas <- rep(0.2, length(color_group_lvls) - length(alpha_select))
ann_geom_values[["alpha"]] <- setNames(c(alpha_select, additional_alphas), color_group_lvls)

# pick additional ggplot parameters
ann_geom_values[["linetype"]] <- setNames(c("solid", "dotted", "dashed"), c("Good", "Poor", "Control"))
ann_geom_values[["coverage_qc_status"]] <- setNames(c(brewer.pal(6, "Set3")[c(1, 4)], "#808080"), c("Good", "Poor", NA))

nextclade_qc_score_lvls <- c("missingData", "mixedSites", "privateMutations", "snpClusters", "stopCodons", "frameShifts")
ann_geom_values[["nc_qc_scores"]] <- setNames(c(colorblind_pal()(8)[3:8]), nextclade_qc_score_lvls)

# create a sample type legend
sample_group_legend <- s %>%
  select(sample_id, ww_group, sample_group, date_facet) %>%
  order_on_other_col(sample_group, sample_id, decreasing = FALSE) %>%
  {
  ggplot(., aes(y = 1, x = sample_id, fill = ww_group)) +
    geom_tile(color = "black") +
    coord_fixed() +
    scale_fill_manual(values = ann_geom_values$sample_type_colors[match(levels(.$ww_group),
                                                                        names(ann_geom_values$sample_type_colors))]) +
    scale_x_discrete(labels = paste0(.$date_facet, "_", .$sample_group)) +
    theme_bw() +
    theme(
      strip.background = element_blank(),
      panel.grid = element_blank(),
      panel.border = element_blank(),
      axis.title = element_blank(),
      axis.text.y = element_blank(),
      axis.text.x = element_text(angle = -30, hjust = 0),
      axis.ticks = element_blank(),
      axis.line = element_blank(),
      plot.margin= margin(0, 0, 0, 0, "pt"),
      legend.position = "bottom"
    ) +
    labs(y = "", x = "", fill = "Samples/Groups") + 
    guides(fill = guide_legend(override.aes = list(size = 2)))
  }

#show_col(ann_geom_values[["sample_type_colors"]])

```

```{r variant colors}

variant_list <- s %>%
  select(nextclade_clade_who, pangolin_lineage, pangolin_note, Nextclade_pango) %>%
  mutate(pangolin_note = gsub("^.*: |; |scorpio.*", "", pangolin_note)) %>%
  mutate(pangolin_note = case_when(pangolin_note == "Assigned from designation hash." ~ NA,
                                   pangolin_note == "" ~ NA,
                                   TRUE ~ pangolin_note)) %>%
    unique() %>%
  separate_rows(pangolin_note, sep=" ") %>%
  pivot_longer(cols = c("pangolin_lineage", "pangolin_note", "Nextclade_pango"), names_to = "variant_num", values_to = "variants") %>%
  mutate(variants = gsub("\\(.*", "", variants)) %>%
  select(-variant_num) %>%
  filter(!is.na(variants)) %>%
  filter(variants!="Unassigned") %>%
  distinct() %>%
  arrange(nextclade_clade_who, variants) %>%
  mutate(combined_list = gsub("\\..*", "", variants))

if(length(unique(variant_list$combined_list)) > 20) {
  stop(simpleError("The number of variants to color is greater than 20. Increase the number of colors in variant_color below"))
}

if(length(unique(variant_list$combined_list)) > 20) {
  stop(simpleError("The number of variants to color is greater than 20. Increase the number of colors in variant_color below"))
}

variant_color <- data.frame(colors = c(calc_pal()(12)[c(6, 2:5, 1, 8:12, 7)],
                                       few_pal(palette = "Light")(8))) %>%
  slice_head(n=length(unique(variant_list$combined_list))) %>%
  bind_cols(lineage = sort(unique(variant_list$combined_list))) %>%
  left_join(variant_list, by = c("lineage" = "combined_list")) %>%
  select(-nextclade_clade_who) %>%
  distinct()
  
var_colorgrad <- NULL
for(var in unique(variant_color$lineage)) {
  subcolors <- variant_color %>%
    filter(lineage == var)

  variant_gradient <- colorRampPalette(c(unique(subcolors$colors),
                                         adjustcolor(unique(subcolors$colors), alpha.f = 0.1)), alpha = TRUE)(nrow(subcolors)) 
    
  subcolors <- data.frame(subcolors = variant_gradient) %>%
    bind_cols(subcolors)
  
  var_colorgrad <- var_colorgrad %>%
    bind_rows(subcolors)
}
  
var_colorgrad<- var_colorgrad %>%
  left_join(variant_list, by = c("variants", "lineage" = "combined_list"))

ann_geom_values[["variants"]] <- var_colorgrad %>%
  select(subcolors, variants) %>%
  distinct() %>%
  pull(subcolors, name = variants) 

ann_geom_values[["parent_variant"]] <- var_colorgrad %>%
  select(colors, lineage) %>%
  distinct() %>%
  pull(colors, name = "lineage")
  
ann_geom_values[["variants"]] <- c("#80808080", "#80808080", ann_geom_values[["variants"]])
names(ann_geom_values[["variants"]]) <- c(NA, "Unassigned", names(ann_geom_values[["variants"]])[-(1:2)])

```

```{r save coverage png, eval=import_data}

dir.create(here("data", "figures"), recursive = TRUE)

SC2_annotation <- data.frame(ORF1a = "266-13468",
                             ORF1b = "13468-21555",
                             S = "21563-25384",
                             ORF3a = "25393-26220",
                             E = "26245-26472",
                             M = "26523-27191",
                             ORF6 = "27202-27387",
                             ORF7a = "27394-27759",
                             ORF7b = "27756-27887",
                             ORF8 = "27894-28259",
                             N = "28274-29533",
                             ORF10 = "29558-29674") %>%
  pivot_longer(cols = everything(), names_to = "gene", values_to = "position") %>%
  mutate(annotation = "SARS-CoV-2")

ddPCR_annotation <- data.frame(N1 = "28287-28358",
                               N2 = "29164-29230") %>%
  pivot_longer(cols = everything(), names_to = "gene", values_to = "position") %>%
  mutate(annotation = "ddPCR")

annotation_legend <- rbind(SC2_annotation, ddPCR_annotation) %>%
  mutate(start = as.numeric(gsub("-.*$", "", position)),
         end = as.numeric(gsub("^.*-", "", position)),
         gene = factor(gene, levels = unique(gene)[order(start, decreasing = FALSE)])) %>%
  ggplot() +
    geom_rect(aes(xmin = start, xmax = end, ymin = 1, ymax = 2, fill = gene), position = position_dodge(width = 0.5)) +
    scale_fill_manual(values = tableau_color_pal(palette = "Tableau 20")(20)) +
    scale_x_continuous(expand = c(0,0), limits = c(0, 30000)) + 
    theme_bw() +
    theme(
      strip.background = element_blank(),
      panel.grid = element_blank(),
      panel.border = element_rect(color = "black", fill = NA, size = 1),
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank(),
      axis.text.x = element_text(size = 8),
      legend.position = "bottom"
    ) +
    labs(x="Genomic position (bp)", y="", fill = "") +
    guides(fill=guide_legend(nrow=1, bycol=TRUE, override.aes = list(size = 0.5)))

slide_window_coverage %>%
  merge(select(s, sample_id, ww_group), by = "sample_id", all.x = TRUE) %>%
  mutate(slide_window_100 = ifelse(slide_window_100 <= 1, 1, slide_window_100)) %>%
  {
  ggplot(., aes(x = position, y = slide_window_100, fill = ww_group, group = sample_id)) +
    geom_area(aes(alpha = ww_group), position = "identity") +
    geom_hline(yintercept = 10, linetype = "dashed", alpha = 0.4) +
    facet_wrap(~ww_group, ncol = 1, scales = "free_y") +
    scale_fill_manual(values = ann_geom_values$sample_type_colors[match(levels(.$ww_group), names(ann_geom_values$sample_type_colors))]) +
    scale_alpha_manual(values = ann_geom_values$alpha[match(levels(.$ww_group), names(ann_geom_values$alpha))]) + 
    scale_x_continuous(expand = c(0,0), limits = c(0, 30000)) + 
    scale_y_continuous(trans = "log10") + 
    theme_bw() +
    theme(
      strip.background = element_blank(),
      panel.grid = element_blank(),
      panel.border = element_rect(color = "black", fill = NA, size = 1),
      legend.position = "none",
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      ) +
    labs(x="", y="Sliding window average (X)") +
  annotation_legend + plot_layout(ncol = 1, heights = c(50, 1))
  } %>%
  ggsave(filename = here("data", "figures", paste0(sequencing_date, "_sliding_coverage_100_bp.png")), plot = ., width = 11, height = 8)

slide_window_coverage %>%
  filter(position >= 21563) %>%
  filter(position <= 25384) %>%
  merge(select(s, sample_id, ww_group), by = "sample_id", all.x = TRUE) %>%
  mutate(slide_window_100 = ifelse(slide_window_100 <= 1, 1, slide_window_100)) %>%
  {
  ggplot(., aes(x = position, y = slide_window_100, fill = ww_group, group = sample_id)) +
    geom_area(aes(alpha = ww_group), position = "identity") +
    geom_hline(yintercept = 10, linetype = "dashed", alpha = 0.4) +
    facet_wrap(~ww_group, ncol = 1, scales = "free_y") +
    scale_fill_manual(values = ann_geom_values$sample_type_colors[match(levels(.$ww_group), names(ann_geom_values$sample_type_colors))]) +
    scale_alpha_manual(values = ann_geom_values$alpha[match(levels(.$ww_group), names(ann_geom_values$alpha))]) + 
    scale_x_continuous(expand = c(0, 50),
                       breaks = c(min(.$position),
                                  labeling::extended(min(.$position),
                                                     max(.$position),
                                                     m = 5),
                                  max(.$position))) + 
    scale_y_continuous(trans = "log10") + 
    theme_bw() +
    theme(
      strip.background = element_blank(),
      panel.grid = element_blank(),
      panel.border = element_rect(color = "black", fill = NA, size = 1),
      legend.position = "none"
      ) +
    labs(x="Genomic position (bp)", y="Sliding window average (X)")
  } %>%
  ggsave(filename = here("data", "figures", paste0(sequencing_date, "_s_gene_sliding_coverage_100_bp.png")), plot = ., width = 11, height = 8)

coverage_loop <- s %>%
  filter(sample_id != "Undetermined") %>%
  filter(!is.na(median_coverage)) %>%
  mutate(loop_group = case_when(grepl("Water control|Reagent control|Mock DNA positive control", sample_group) ~ "Sequencing-control",
                                TRUE ~ sample_group)) %>%
  select(loop_group) %>%
  unique() %>%
  pull()

date_loop <- s %>%
  filter(sample_id != "Undetermined") %>%
  filter(!is.na(median_coverage)) %>% 
  select(!!date_col) %>%
  unique() %>%
  pull()
  
  
for(sites in coverage_loop) {
  for (dates in as.character(date_loop)) {
  filtered_samp_name <- s %>%
    mutate(loop_group = case_when(grepl("Water control|Reagent control|Mock DNA positive control", sample_group) ~ "Sequencing-control",
                                  TRUE ~ sample_group),
           date_char = as.character(!!date_col)) %>%
    filter(sample_id != "Undetermined") %>%
    filter(loop_group %in% sites) %>%
    filter(date_char %in% dates) %>%
    mutate(sample_name = case_when(isControl ~ sample_group,
                                   TRUE ~ gsub("^Test-|^WW-", "", sample_name))) %>%
    mutate(coverage_qc_status = case_when(is.na(coverage_qc_status) & !isControl ~ "Poor",
                                          is.na(coverage_qc_status) & isControl ~ "Control",
                                          TRUE ~ coverage_qc_status)) %>%
    mutate(coverage_qc_status = factor(coverage_qc_status, levels = names(ann_geom_values$linetype))) %>%
    mutate(parent_variant = case_when(is.na(Nextclade_pango) & !isControl ~ "no results",
                                 is.na(Nextclade_pango) & isControl ~ "control",
                                 TRUE ~ gsub("^([A-Za-z]*\\.[0-9]*).*", "\\1", Nextclade_pango))) %>%
    mutate(parent_variant = factor(parent_variant, levels = c(names(ann_geom_values$parent_variant), "no results", "control"))) %>%
    droplevels() %>%
    select(sample_id, sample_name, parent_variant, coverage_qc_status)
  
  if (nrow(filtered_samp_name) == 0) {
    next
  }
  
  sample_name_label <- setNames(as.character(filtered_samp_name$sample_name), filtered_samp_name$sample_id)
  
    empty_files <- slide_window_coverage %>%
      filter(sample_id %in% filtered_samp_name$sample_id)
    
    if (nrow(empty_files) == 0) {
      next
    }

  slide_window_coverage %>%
    filter(sample_id %in% filtered_samp_name$sample_id) %>%
    mutate(slide_window_100 = ifelse(slide_window_100 > 10, 10, slide_window_100)) %>%
    merge(filtered_samp_name, by = "sample_id", all = TRUE) %>%
    droplevels() %>%
    {
    ggplot(., aes(x = position, y = slide_window_100, group = sample_id, fill = parent_variant)) +
      geom_line(aes(linetype = coverage_qc_status), color = "black") +
      geom_area(position = "identity") +
      facet_wrap(~sample_id, ncol = 1, scales = "free_y", labeller = labeller(sample_id = sample_name_label)) +
      scale_linetype_manual(values = ann_geom_values$linetype[match(levels(.$coverage_qc_status), names(ann_geom_values$linetype))]) +
      scale_fill_manual(values = c(ann_geom_values$parent_variant, setNames(c("#808080", "#000000"), c("no results", "control")))[match(levels(.$parent_variant), c(names(ann_geom_values$parent_variant), "no results", "control"))]) +
      scale_x_continuous(expand = c(0,0), limits = c(0, 30000)) + 
      theme_bw() +
      theme(
        strip.background = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_rect(color = "black", fill = NA, size = 1),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        ) +
      labs(x=paste0(sites, " samples"),
           y="Sliding window average (X)",
           fill = "Parent\nvariant",
           linetype = "Coverage Quality") +
      guides(linetype = guide_legend(order = 1), fill = guide_legend(order = 2)) +
    annotation_legend + plot_layout(ncol = 1, heights = c(50, 1))
    } %>%
    ggsave(filename = here("data", "figures", paste0(sequencing_date, "_sample_sliding_coverage_100_bp_site_", dates, "_", sites, ".png")), plot = ., width = 11, height = 8)
  }
}
```

```{r empty plate to plot}

empty_plate <- data.frame(plate_row = unlist(lapply(LETTERS[1:8], function(x) rep(x, 12))), plate_col = sprintf("%02d", rep(1:12, 8)), plate = 1) %>%
  mutate(plate_coord = paste0(plate, "_", plate_row, plate_col))

```

