---
title: |
    | 
    | 
    | 
    | 
    | ![](../aux_files/pdf_generation/forKnitting/logo_blk.png){width=4.5in}  
    | 
    | Sequencing Lab
    | 
    | QC report for sequencing run
    | 
    | `r basename(here())`

author: "Philadelphia Public Health Laboratory"
date: \today
geometry: margin=3cm
output:
    pdf_document:
        template: ../aux_files/pdf_generation/forKnitting/toc_after.tex
        keep_tex: false
        toc: true
        toc_depth: 3
        includes:
            in_header: ../aux_files/pdf_generation/forKnitting/TeX_packages_commands.sty

---

\newpage

\tableofcontents

```{r settings for processing the sequencing run, include=FALSE, eval=FALSE}

#############################################################
# 1_archive_n_upload_run.R (Sequencer and BaseSpace settings) 
#############################################################

sequencer_select <- 2 # set variable as 1 for MiSeq or 2 for NextSeq

run_uploaded_2_basespace <- FALSE # set this to TRUE if the run was uploaded to BaseSpace from the sequencer directly

sample_type_acronym <- "WW"

pathogen_acronym <- "SC2"

# set show_col_types to FALSE
options("readr.show_col_types" = FALSE)

#################################################################
# 2_generate_metadata_n_assign_barcodes.R (Sample sheet settings)
#################################################################

remove_sample_from_bcl_samplesheet <- c("") #add in sample names to remove from demultiplexing

sample_w_empty_reads <- c("") #add in sample ids that have empty fastq files

prj_description <- "COVIDSeq" #no spaces, should be the same as the R project

if(nchar(prj_description) > 25) {
  stop(simpleError("prj_description is too long. Needs to be 25 chars or less"))
}

####################################
# 3_run_nextflow.R (Cecret settings)
####################################

#tag of nextclade dataset to use for lineage assignment
#leave empty to use the latest nextclade dataset
nextclade_dataset_version <- ""

update_freyja_and_cecret_pipeline <- FALSE

cecret_version <- "3.15.24324"

remove_undetermined_file <- FALSE

```

```{r archive and upload sequencing run data, message = TRUE, include=FALSE, eval = FALSE}

#this script compresses the sequencing run folder and uploads it to S3
source("code/1_archive_n_upload_run.R")

```

```{r generate sequencing metadata and sample sheets, message = TRUE, include=FALSE, eval = FALSE}

#this script generates the SampleSheet.csv file required for demultiplexing the sequencing run
#the required file is a sequencing_metadata_sheet.xlsx file filled out by the sequencing wet lab scientists with the sample barcode information
#this file should be placed in metadata/munge
source("code/2_generate_metadata_n_assign_barcodes.R")

```

```{r run nextflow, message = TRUE, include=FALSE, eval = FALSE}

#this script submits the nextflow jobs for demultiplexing and running the cecret pipeline on the ec2 instance
source("code/3_run_nextflow.R")

```

```{r setup for the pdf, echo=FALSE, warning=FALSE}

library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=FALSE,
  cache.lazy = FALSE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE, #Set message to NA in chunks where message should be sent to stderr
  dpi=100,
  fig.width=6,
  fig.height=6,
  dev.args = list(pdf = list(useDingbats = FALSE))
)

#this lets you loop through variables and produce multiple pander tables and ggplots in a single code block!
#this also means that you have to put results='asis' in any block that has pander output
pander::panderOptions("knitr.auto.asis", FALSE)

```

```{r load data, child = 'code/preamble.Rmd', eval = load_preamble}

```

```{r make figures, child = 'code/data_analysis.Rmd'}

```

```{r generate time stamped report, message=TRUE, include=FALSE, eval=FALSE}
#generate the pdf report

### =============
### Manual inputs
### =============

# need to import the data into a single tsv for the first time?
import_data <- TRUE
report_only_cdc_lineages <- TRUE

### =================
### End manual inputs
### =================

rstudioapi::documentSaveAll()
library(here)
dir.create(here("output"))
project_name <- basename(here())
options("readr.show_col_types" = FALSE)

qc_report_fp <- paste0(project_name, "_QC_Report.Rmd")
qc_output_fp <- here("output", paste0(project_name, ".QC.report_gen.on.", Sys.Date(), ".pdf"))

# load preamble the first time?
load_preamble <- TRUE

# print table as kable for pdf?
use_kable <- TRUE
rmarkdown::render(qc_report_fp, output_file = qc_output_fp)

load_preamble <- FALSE
use_kable <- FALSE
rmarkdown::render(qc_report_fp, output_file = "README.md",
                 output_format = rmarkdown::md_document(variant = "gfm", toc = TRUE))

```

```{r save pdf shared drive, include=FALSE, eval=FALSE}

rstudioapi::documentSaveAll()
library(here)
library(magrittr)

#this file needs to sit in a [aux_files/config] directory path above this project directory
tryCatch(
  {
    source(file.path(dirname(here()), "aux_files", "r_scripts", "config", "config_variables.R"))
  },
  error = function(e) {
  stop (simpleError("The config_variables.R file needs to sit in a [aux_files/r_scripts/config] directory path above this project directory"))
  }
)

project_name <- basename(here())

copy_to <- file.path(shared_drive_fp, "Sequencing_results", "COVIDSeq", "wastewater")
dir.create(file.path(copy_to, project_name), recursive = TRUE)

pdf_fp <- here("output", list.files(here("output"), pattern = ".pdf"))
for(i in pdf_fp){
  file.copy(i, file.path(copy_to, project_name), overwrite = TRUE)
}

file.copy(here("upload", "epi"), file.path(copy_to, project_name), recursive = TRUE, overwrite = TRUE)

file.copy(here("upload", "lims"), file.path(copy_to, project_name), recursive = TRUE, overwrite = TRUE)

```

```{r GISAID BioSample SRA submission, message=TRUE, include=FALSE, eval=FALSE}

#this script submits the data to NCBI and GISAID

#do a test upload first. change this to FALSE when SRA upload and gisaid upload testing is successful and you want to submit the files for real
test_upload <- TRUE

#the PHL2 sample_id of samples to remove from the merged FASTA sample
#should be in the format of PHL2-[sequencer]-[plate position]-[sequencing date]
failed_sample_fasta <- c("")

source("code/4_upload_data_2_public_repo.R")

```
