
```{r libraries, message=FALSE, warning=FALSE}
library(here)
library(tidyverse)
library(rvest)
library(pander)
library(kableExtra)
library(ggsci)
library(RColorBrewer)
library(patchwork)
library(reticulate)

```

```{r versions and dates manual inputs}

#file path of the parent directory holding the sequencing run folders
sequencing_folder_fp <- "Z:/MiSeq_runs/COVIDSeq"

#full file path of the git shell executable
sh_exe_fp <- "C:/Users/vincent.tu/AppData/Local/Programs/Git/bin/sh.exe"

#application to run on basespace
app2run <- "DRAGEN COVID Lineage"

primer_select <- 2 # set variable as 1 for V3, 2 for V4, or 3 for V4.1

artic_primer_scheme <- c("ncov2019", "articv4", "articv4.1")[primer_select]
bs_bed_file_used <- c("nCov-2019.bed", "artic_V4_primer_scheme_NC_045512.2.bed", "artic_v4.1_primers_nc045512.2.bed")[primer_select]

# minimum acceptable number of reads
min_reads <- 1000000

auth_token <- list.files(file.path(dirname(here()), "aux_files", "seqsender"), pattern = "gisaid_uploader.authtoken", full.names = TRUE)
seqsender_config_fp <- list.files(file.path(dirname(here()), "aux_files", "seqsender"), pattern = "yaml", full.names = TRUE)

```

```{r functions and load signature}

#this file needs to sit in a [aux_files/functions] directory path above this project directory
tryCatch(
  {
    source(file.path(dirname(here()), "aux_files", "functions", "R_all_functions_v3.R"))
  },
  error = function(e) {
  stop (simpleError("The R_all_functions_v3.R file needs to sit in a [aux_files/functions] directory path above this project directory"))
  }
)

#this file needs to sit in a [aux_files/signature] directory path above this project directory
tryCatch(
  {
    signature_fp <- list.files(file.path(dirname(here()), "aux_files", "signature"), pattern = ".png|.jpeg|.tif*|.gif", full.names = TRUE)
    tag <- signature_fp[!grepl("blank.png", signature_fp)]
    blank <- signature_fp[grepl("blank.png", signature_fp)]
  },
  error = function(e) {
  stop (simpleError("The signature image file needs to sit in a [aux_files/signature] directory path above this project directory"))
  }
)

```

```{r find newest run folder}

project_name <- basename(here())
sequencing_date <- as.Date(gsub("_.*", "", project_name))

#get the newly added run folder
run_folder <- sequencing_folder_fp %>%
  list.files(full.names = T) %>%
  file.info() %>%
  filter(grepl(format(sequencing_date, "%y%m%d"), rownames(.))) %>%
  rownames()

folder_date <- paste0("20", gsub(".*/|_.*", "", run_folder)) %>%
  as.Date(format = "%Y%m%d")

if(folder_date != gsub("_.*", "", basename(here()))) {
 stop(simpleError("The run date on the sequencing folder does not match the date of this RStudio project!"))
}

```

```{r copy run folder files}

dir.create(here("data", "demux_reads"), recursive = TRUE)

latest_demux <- max(list.files(run_folder, pattern = "Alignment_"))

#copy from
cp_list <- c(list.files(here(run_folder, latest_demux), pattern = "CompletedJobInfo.xml", recursive = TRUE, full.names = TRUE),
             list.files(here(run_folder, latest_demux), pattern = "DemultiplexSummaryF1L1.txt", recursive = TRUE, full.names = TRUE),
             list.files(here(run_folder, latest_demux), pattern = "FastqSummaryF1L1.txt", recursive = TRUE, full.names = TRUE),
             list.files(here(run_folder, latest_demux), pattern = "SampleSheetUsed.csv", recursive = TRUE, full.names = TRUE))

#copy to
invisible(sapply(cp_list, file.copy, here("data", "demux_reads"), TRUE))

```

```{r file paths for metadata and reads}

### mapping file path
#try to get it automagically
mapping_file_fp <- list.files(here("metadata"), pattern = ".tsv|.csv", full.names = TRUE)

### demultiplexing results from BaseSpace
gen_fq_fp <- here("data", "demux_reads", "CompletedJobInfo.xml")
demuxsum_fp <- here("data", "demux_reads", "DemultiplexSummaryF1L1.txt")
fastqsum_fp <- here("data", "demux_reads", "FastqSummaryF1L1.txt")
SampleSheet_fp <- here("data", "demux_reads", "SampleSheetUsed.csv")

```

```{r load bcl2fq, eval = FALSE}

bcl2fq_version <- read_delim(b2fq_v_fp, skip = 1, col_names = FALSE) %>%
  slice(1) %>%
  select(X2) %>%
  pull()

#just reading in command file as one column
bcl2fq_flags <- read_delim(b2fq_c_fp, col_names = FALSE, col_types = cols(.default = "c"), delim = ";") %>%
  slice(1) %>%
  mutate(X1 = gsub("^.*bcl2fastq", "bcl2fastq", X1)) %>%
  mutate(X1 = gsub(" --", "\n\t--", X1)) %>%
  mutate(X1 = gsub(" -", "\n\t-", X1)) %>%
  pull()

```

```{r load GenerateFASTQ module}

JobCompletedInfo <- xml2::read_xml(gen_fq_fp)

workflow_name <- JobCompletedInfo %>%
  html_elements("Analysis") %>%
  html_text2()

gen_fq_version <- JobCompletedInfo %>%
  html_elements("WorkflowVersion") %>%
  html_text2()

```

```{r load read numbers}

# read counts by sample number
fastq_sum <- read_delim(fastqsum_fp) %>%
  group_by(SampleNumber) %>%
  #PF means reads that have passed filter
  summarise(total_raw_reads = sum(NumberOfReadsRaw)*2,
            read_counts = sum(NumberOfReadsPF)*2) %>%
  ungroup()

# SampleSheet.csv loading

sample_sheet <- load_sample_sheet(SampleSheet_fp)

samp_sh_header <- data.frame(X1 = unlist(sample_sheet$Header)) %>%
  mutate(col_names = gsub(",.*", "", X1)) %>%
  mutate(col_names = gsub(" ", "_", col_names)) %>%
  mutate(X1 = gsub(".*,", "", X1)) %>%
  pivot_wider(names_from = "col_names", values_from = "X1")

samp_sh_check <- samp_sh_s <- data.frame(X1 = unlist(sample_sheet$Data), stringsAsFactors = FALSE) %>%
  #split each line into columns using row 1 as the column names
  separate(col = "X1", strsplit(as.character(slice(., 1)), ",")[[1]], sep = ",") %>%
  slice(-1) %>%
  rename(sample_id = "Sample_ID") %>%
  mutate(I7_ID = I7_Index_ID == UDI_Index_ID) %>%
  mutate(I5_ID = I5_Index_ID == UDI_Index_ID) %>%
  select(UDI_Index_ID, I7_ID, I5_ID)

#if any of the barcode ID's dont match, raise an error
if(!all(c(samp_sh_check$I7_ID, samp_sh_check$I7_ID))) {
  stop(simpleError("One of the barcode ID's don't match! Check samp_sh_check dataframe"))
}

samp_sh_s <- data.frame(X1 = unlist(sample_sheet$Data), stringsAsFactors = FALSE) %>%
  #split each line into columns using row 1 as the column names
  separate(col = "X1", strsplit(as.character(slice(., 1)), ",")[[1]], sep = ",") %>%
  slice(-1) %>%
  #this is for bcl2fastq
  #rename(SampleName = "Sample_Name", SampleNumber = "BASESPACE_ONLY_ORIGINAL_SAMPLE_NUMBER") %>%
  #mutate(Barcode_sequence = paste0(I7_Index_ID, "+", I5_Index_ID)) %>%
  #select(SampleName, I7_Index_ID, I5_Index_ID, Barcode_sequence, Sample_Type, UDI_Index_ID)
  rename(sample_id = "Sample_ID") %>%
  mutate(I7_Index_ID = index, I5_Index_ID = index2) %>%
  mutate(Barcode_sequence = paste0(I7_Index_ID, "+", I5_Index_ID)) %>%
  select(sample_id, I7_Index_ID, I5_Index_ID, Barcode_sequence, UDI_Index_ID) %>%
  mutate(sample_id = gsub("_", "-", sample_id))

#merge sample names with read counts
samp_reads <- read_delim(demuxsum_fp, col_names = FALSE) %>%
  filter(grepl("Sample", X1)) %>%
  t() %>%
  as.data.frame(stringsAsFactors = FALSE) %>%
  purrr::set_names(as.character(slice(., 1))) %>%
  slice(-1) %>%
  filter(!is.na(SampleNumber)) %>%
  merge(fastq_sum, by = "SampleNumber", all = TRUE) %>%
  rename(sample_id = "SampleName") %>%
  mutate(sample_id = gsub("_", "-", sample_id)) %>%
  mutate(sample_id = ifelse(sample_id == "None", "Undetermined", sample_id)) %>%
  merge(samp_sh_s, by = "sample_id", all = TRUE) %>%
  mutate(percent_reads_passed_filter = read_counts/total_raw_reads) %>%
  mutate(percent_of_lane = read_counts/sum(read_counts))


top_unkwn_indices <- read_delim(demuxsum_fp, col_names = FALSE) %>%
  filter(is.na(X6)) %>%
  filter(!is.na(X4)) %>%
  rename(index_sequence1 = "X1", index_sequence2 = "X3", number_of_reads = "X5") %>%
  filter(!grepl("###", index_sequence1)) %>%
  select(index_sequence1, index_sequence2, number_of_reads) %>%
  mutate(index_sequence = paste0(index_sequence1, "+", index_sequence2)) %>%
  select(-c(index_sequence1, index_sequence2)) %>%
  mutate(number_of_reads = as.numeric(number_of_reads)) %>%
  order_on_other_col(index_sequence, number_of_reads) %>%
  slice(1:10) %>%
  droplevels()

```

```{r upload data to basespace and run the app, message=FALSE, eval = dl_bs_data}

dir.create(here("data", "processed_bs"), recursive = TRUE)

list_of_indexes <- list.files(here(run_folder, latest_demux), pattern = "*I[12]_001.fastq.gz", recursive = TRUE, full.names = TRUE)

#if there are index files in the run folder, move them somewhere else
if(length(list_of_indexes) > 0) {
  
  #make index folder by replace the slashes
  dir.create(file.path(dirname(dirname(list_of_indexes[1])), "indices"), recursive = TRUE)
  
  for(index in list_of_indexes) {
    file.rename(index, file.path(dirname(dirname(list_of_indexes[1])), "indices", basename(index)))
  }

}

fastq_folder <- dirname(list.files(here(run_folder, latest_demux), pattern = "R[12]_001.fastq.gz", recursive = TRUE, full.names = TRUE)[1])

#have to break the shell script into parts because of proxy authentication issue which can be solved by initializing a new shell script each time

### 1 - create the basespace project
bs_cli_create_fp <- file.path(dirname(here()), "aux_files", "bs_cli_commands", "bs_cli_1_create_project.sh")
project_id <- bs_cli_submit(sh_exe_fp, bs_cli_create_fp, project_name)
write(paste0(c("project_id", project_id), collapse = ","), here("data", "bs_metadata.csv"))

### 2 - upload the fastq files
bs_cli_upload_fp <- file.path(dirname(here()), "aux_files", "bs_cli_commands", "bs_cli_2_upload_fastq.sh")
files_upload <- bs_cli_submit(sh_exe_fp, bs_cli_upload_fp, c(project_id,
                                             shQuote(fastq_folder, type = "sh")))

### 3 - get the latest app version
bs_cli_app_ver_fp <- file.path(dirname(here()), "aux_files", "bs_cli_commands", "bs_cli_3_get_app_version.sh")
dragen_app_version <- bs_cli_submit(sh_exe_fp, bs_cli_app_ver_fp, shQuote(app2run, type = "sh"))
write(paste0(c("dragen_app_version", dragen_app_version), collapse = ","), here("data", "bs_metadata.csv"), append = TRUE)


### 4 - launch the app
bs_cli_launch_fp <- file.path(dirname(here()), "aux_files", "bs_cli_commands", "bs_cli_4_launch_app.sh")
app_session_id <- bs_cli_submit(sh_exe_fp, bs_cli_launch_fp, c(shQuote(app2run, type = "sh"),
                                                              dragen_app_version,
                                                              project_id,
                                                              artic_primer_scheme))
write(paste0(c("app_session_id", app_session_id), collapse = ","), here("data", "bs_metadata.csv"), append = TRUE)

### 5 - wait until the app finished running
bs_cli_await_fp <- file.path(dirname(here()), "aux_files", "bs_cli_commands", "bs_cli_5_await.sh")
awaiting <- bs_cli_submit(sh_exe_fp, bs_cli_await_fp, app_session_id)

### 6 - download data
bs_cli_download_fp <- file.path(dirname(here()), "aux_files", "bs_cli_commands", "bs_cli_6_download.sh")
downloading <- bs_cli_submit(sh_exe_fp, bs_cli_download_fp, c(project_id,
                                              shQuote(here("data", "processed_bs"), type = "sh")))

### 7 - get app date
bs_cli_get_date_fp <- file.path(dirname(here()), "aux_files", "bs_cli_commands", "bs_cli_7_get_date.sh")
app_run_date <- bs_cli_submit(sh_exe_fp, bs_cli_get_date_fp, app_session_id) %>%
  as.Date() %>%
  as.character()
write(paste0(c("app_run_date", app_run_date), collapse = ","), here("data", "bs_metadata.csv"), append = TRUE)

```

```{r get nextclade version by downloading Undetermined file, eval = import_data}

nextclade_version_fp <- list.files(here("data", "processed_bs"), pattern = "Undetermined_report_metrics.json", recursive = TRUE, full.names = TRUE)

#reading in json like a csv and then handling it in a hacky way
nc_v <- nextclade_version_fp %>%
  read_csv(col_names = FALSE) %>%
  select_if(~ any(grepl("nextclade_version", .))) %>%
  mutate_all(~gsub(".*nextclade ", "", .)) %>%
  pull()

```

```{r unzip fastq files and bed files, eval = import_data}

#find all zipped files
zip_fps <- list.files(here("data", "processed_bs"), pattern = ".zip", recursive = TRUE, full.names = TRUE)

#extract scrubbed fastq files
scrubbed_zip_files <- zip_fps[grepl("ScrubbedFASTQsNode", zip_fps)]
invisible(sapply(scrubbed_zip_files, function(x) unzip(x, exdir = dirname(x))))

#extract just the bed files
bed_zip_files <- zip_fps[grepl("_all_output_files", zip_fps)]
#sometimes these files are bed and other times they are tsv files...
#be wary of the extract to file path. if the length of the path exceeds 256 characters, it may throw an error
invisible(sapply(bed_zip_files, function(x) unzip(x, exdir = here("data", "processed_bs"), files = as.character(paste0("consensus/", gsub(".*/|_all_output_files.zip$", "", x), ".consensus_coverage_from_filtered_bam.tsv")))))

invisible(sapply(bed_zip_files, function(x) unzip(x, exdir = here("data", "processed_bs"), files = paste0("consensus/", gsub(".*/|_all_output_files.zip$", "", x), ".consensus_coverage_from_filtered_bam.bed"))))


```

```{r load coverage results, eval = import_data}

#this bed file can only be found in the zipped output files
cov_files_fps <- list.files(here("data", "processed_bs", "consensus"), pattern = "consensus_coverage_from_filtered_bam", recursive = TRUE, full.names = TRUE)

coverage <- cov_files_fps %>%
  data_frame(FileName = .) %>%
  group_by(FileName) %>%
  do(read_delim(.$FileName, col_names = FALSE)) %>%
  ungroup() %>%
  filter(X1 == "NC_045512.2") %>%
  #get column 3 if its a tsv, otherwise get column 4
  mutate(map_per_position = ifelse(grepl("tsv$", FileName), X3, X4)) %>%
  select(FileName, map_per_position) %>%
  mutate(sample_id = gsub(".*/", "", FileName)) %>%
  mutate(sample_id = gsub("\\.consensus.*", "", sample_id)) %>%
  select(-c(FileName)) %>%
  group_by(sample_id) %>%
  summarise(pct_genome_coverage_over_30x = sum(map_per_position >= 30)/n(),
            median_coverage = median(map_per_position)) %>%
  ungroup()

```

```{r load kmer results, eval = import_data}

kmer_files_fps <- list.files(here("data", "processed_bs"), pattern = ".detect.report.csv", recursive = TRUE, full.names = TRUE)

kmer <- kmer_files_fps %>%
  data_frame(FileName = .) %>%
  group_by(FileName) %>%
  do(read_delim(.$FileName, col_names = FALSE)) %>%
  ungroup() %>%
  select(-c(FileName, X1)) %>%
  pivot_wider(names_from = "X3", values_from = "X4") %>%
  rename_with(~str_replace(., "SARS-CoV2", "SARS-CoV-2")) %>%
  rename(sample_id = "X2", `SARS-CoV-2 Total Reference Kmers` = "SARS-CoV-2 Kmers", `SARS-CoV-2 Targets found` = "SARS-CoV-2 Targets") %>%
  mutate(`SARS-CoV-2` = ifelse(`SARS-CoV-2` == "+", "Positive", "Negative")) %>%
  mutate(`SARS-CoV-2` = factor(`SARS-CoV-2`, levels = c("Positive", "Negative"))) %>%
  mutate(across(!sample_id & !`SARS-CoV-2`, as.numeric))

```

```{r load Pangolin results, eval = import_data}

pango_files_fps <- list.files(here("data", "processed_bs"), pattern = ".lineage_report.csv", recursive = TRUE, full.names = TRUE)

pango <- pango_files_fps %>%
  data_frame(FileName = .) %>%
  group_by(FileName) %>%
  do(read_delim(.$FileName, col_names = TRUE)) %>%
  ungroup() %>%
  select(-c(FileName)) %>%
  rename(pango_qc_status = "qc_status", sample_id = "taxon") %>%
  mutate(scorpio_lineage = gsub(".* \\(|-like.*|Unassigned.*", "", scorpio_call)) %>%
  mutate(scorpio_call = gsub(" \\(.*", "", scorpio_call)) %>%
  mutate(scorpio_call = gsub("Probable ", "", scorpio_call)) %>%
  mutate(scorpio_call = ifelse(is.na(scorpio_call) & lineage == "B", "Reference", scorpio_call))

```

```{r load NextClade results, eval = import_data}

nc_files_fps <- list.files(here("data", "processed_bs"), pattern = ".nextclade.tsv", recursive = TRUE, full.names = TRUE)

next_clade <- nc_files_fps %>%
  data_frame(FileName = .) %>%
  group_by(FileName) %>%
  do(read_delim(.$FileName, col_names = TRUE, col_types = cols(missing = col_character()))) %>%
  ungroup() %>%
  select(-c(FileName)) %>%
  rename(nc_qc_status = "qc.overallStatus", sample_id = "seqName") %>%
  mutate(nc_VOC = gsub(".*\\(|\\).*", "", clade)) %>%
  mutate(nc_VOC = gsub("^19A$", "Reference", nc_VOC))

```


```{r write bs_results, eval = import_data}

bs_meta <- read_csv(here("data", "bs_metadata.csv"), col_names = FALSE) %>%
  t() %>%
  `colnames<-`(.[1, ]) %>%
  as.data.frame() %>%
  slice(-1)

merge(coverage, kmer, by = "sample_id", all = TRUE) %>%
  merge(pango, by = "sample_id", all = TRUE) %>%
  merge(next_clade, by = "sample_id", all = TRUE) %>%
  mutate(date_pipeline_was_run = as.Date(bs_meta$app_run_date)) %>%
  mutate(dragen_covid_lineage_version = as.character(bs_meta$dragen_app_version)) %>%
  mutate(nextclade_version = nc_v) %>%
  write_delim(here("data", "bs_results.tsv"), delim = "\t")

```

```{r load data}

bs_results <- read_tsv(here("data", "bs_results.tsv"))

pangolin_software_version <- as.character(unique(bs_results[!is.na(bs_results$pangolin_version), "pangolin_version"]))

lineage_assignment_software_version <- paste0(unique(bs_results$version)[!is.na(unique(bs_results$version))], collapse = "/")

scorpio_software_version <- paste0(unique(bs_results$scorpio_version)[!is.na(unique(bs_results$scorpio_version))], collapse = "/")

constellation_version <- gsub("^v", "", paste0(unique(bs_results$constellation_version)[!is.na(unique(bs_results$constellation_version))], collapse = "/"))

nextclade_version <- unique(bs_results$nextclade_version)

date_pipeline_was_run <- unique(bs_results$date_pipeline_was_run)

dragen_covid_lineage_version <- unique(bs_results$dragen_covid_lineage_version)

```

```{r sample_sheet_import and columns, echo=FALSE}

ctrl_regex <- "EBneg|Extract|Vibriolambda|Blank|MockDNA|DNAfreewater|Geneblock|Emptywell|Control|POS|NEG|PC|NC"

sample_type_fct <- c("Unassigned reads", "Empty well control", "Buffer control", "Mock DNA positive control")

s_meta <- read_delim(mapping_file_fp, delim = ",")

s <- s_meta %>%
  merge(samp_reads, by = c("sample_id", "UDI_Index_ID"), all = TRUE) %>%
  merge(bs_results, by = "sample_id", all = TRUE) %>%
  mutate(read_counts = ifelse(is.na(read_counts), 0, read_counts), total_raw_reads = ifelse(is.na(total_raw_reads), 0, total_raw_reads)) %>%
  mutate(isControl = grepl(ctrl_regex, sample_id, ignore.case = TRUE)|grepl(ctrl_regex, sample_type, ignore.case = TRUE)) %>%
  #make a new sample type based on if there were pangolin and nextclade results. DRAGEN App uses the <90 on the COVID targets as their threshold
  #Based on their manual, targets seem to be "amplicons" or regions of longer stretches of DNA that are unique to COVID (longer than their 32 k-mers that they use for k-mer detection)
  mutate(sample_type = case_when(grepl("^Undetermined$", sample_id) ~ sample_type_fct[1],
                                 ((`SARS-CoV-2 Targets found` < 90) & !isControl) ~ paste0(sample_type, " - No Variant Calls"),
                                 TRUE ~ sample_type)) %>%
  #rearrange the sampletypes by ctrls first
  mutate(sample_type = factor(sample_type, levels = c(sample_type_fct,
                                                      sort(unique(sample_type)[!grepl(paste0(sample_type_fct, collapse = "|"), unique(sample_type))])))) %>%
  droplevels()

color_by <- NULL
shape_by <- NULL
quality_by <- "sample_type"
potential_headers <- c("study_group", "study_day", "subject_id",
                       "current_antibiotics", "host_species", "cage_number") #pick 2
header_idx <- which(is.element(potential_headers, colnames(s)))

if(length(header_idx)>0){
  color_by <- potential_headers[header_idx[1]]
}
if(length(header_idx)>1){
  shape_by <- potential_headers[header_idx[2]]
}

show.text <- nrow(s) > 40
```

```{r for basic report}

has_CT <- any(grepl("^CT$", colnames(s)))

has_RLU <- any(grepl("^RLU$", colnames(s)))

### ====================================
### Select specific groups to plot for QC
### ====================================
s_toPlot <- s %>%
  filter(!grepl("None|Undetermined", sample_id)) %>%
  filter(!grepl(" - No Variant Calls$", sample_type)) %>%
  #filter out samples without any results
  filter(pango_qc_status != "fail") %>%
  mutate(sample_id = as.character(sample_id)) %>%
  droplevels()

```

```{r sample type attributes}

lvls <- c(levels(s$sample_type), NA)

ann_geom_values <- list()

#leave Unassigned reads as purple
ann_geom_values[["sample_type_colors"]] <- setNames(pal_d3("category20")(20)[c(15, 19, 16, 2, 1, 4, 8)], lvls)

#make the controls and unassigned reads into triangle and diamond shapes
ann_geom_values[["sample_type_shapes"]] <- setNames(c(24:25, 8, 23,
                                                      21:22,
                                                    c(9:14)[1:(length(lvls)-6)]), lvls)

#make the controls and unassigned reads samples bigger in size
ann_geom_values[["sample_type_sizes"]] <- setNames(c(4, 4, 4, 4, rep(1, length(lvls)-3)), lvls)

#scales::show_col(ann_geom_values[["sample_type_colors"]])

```

```{r variant colors}

i <- 1
ann_geom_values[["variants"]] <- c()
ann_geom_values[["who_label"]] <- c()

for(variant in sort(unique(s$scorpio_call)[!is.na(unique(s$scorpio_call))])) {
  
  #only coloring first two variants
  variant_list <- s %>%
    filter(scorpio_call == variant) %>%
    #is_designated means that the sample was assigned to a reference strain exactly using PANGO algorithm (decision tree) rather than PUSHER
    #the note section here for PUSHER will be 'Assigned from designation hash.' so its useless
    #nextclade VOC shouldn't be too different from scorpio VOC
    select(note, Nextclade_pango, scorpio_lineage) %>%
    mutate(note = gsub("^.*: |; scorpio.*", "", note)) %>%
    mutate(note = ifelse(note == "Assigned from designation hash.", NA, note)) %>%
    separate(note, into = c("pusher_variant1", "pusher_variant2"), sep = " ", extra = "merge") %>%
    pivot_longer(cols = c("pusher_variant1", "pusher_variant2", "Nextclade_pango", "scorpio_lineage"), names_to = "variant_num", values_to = "variants") %>%
    mutate(variants = gsub("\\(.*", "", variants)) %>%
    select(variants) %>%
    filter(!is.na(variants)) %>%
    unique() %>%
    pull() %>%
    str_sort(numeric = TRUE, decreasing = TRUE)
  
  variant_color <- brewer.pal(8, "Dark2")[i]
  
  variant_gradient <- colorRampPalette(c(variant_color, adjustcolor(variant_color, alpha.f = 0.1)), alpha = TRUE)(length(variant_list))
  names(variant_gradient) <- variant_list
  
  who_color <- variant_color
  names(who_color) <- variant
  
  ann_geom_values[["variants"]] <- c(ann_geom_values[["variants"]], variant_gradient)
  ann_geom_values[["who_label"]] <- c(ann_geom_values[["who_label"]], who_color)
  
  i <- i + 1
  
}

ann_geom_values[["variants"]] <- c("#80808080", ann_geom_values[["variants"]])
names(ann_geom_values[["variants"]]) <- c(NA, names(ann_geom_values[["variants"]])[-1])

#scales::show_col(ann_geom_values[["variants"]])

who_legend <- data.frame(x = 1, y = 1, fill_col = names(ann_geom_values[["who_label"]])) %>%
  ggplot(aes(x = x, y = y, fill = fill_col)) +
    geom_bar(stat="identity", position = position_fill(reverse = TRUE), alpha = 0) +
    scale_fill_manual(values = ann_geom_values$who_label) +
    guides(fill = guide_legend(override.aes = list(alpha=1), reverse=T, order = 1)) + 
    theme_bw() +
    theme(
      strip.text = element_text(size = 12),
      strip.background = element_blank(),
      panel.grid = element_blank(),
      panel.border = element_blank(),
      axis.title = element_blank(),
      axis.text = element_blank(),
      axis.ticks = element_blank(),
      legend.position = c(0.5, 0.5), # move the legend to the center
      plot.margin=unit(c(0, 0, 0, 0), "cm")
    ) +
    labs(fill = "WHO label", y = "", x = "")

```

```{r empty plate to plot}

empty_plate <- data.frame(plate_row = unlist(lapply(LETTERS[1:8], function(x) rep(x, 12))), plate_col = sprintf("%02d", rep(1:12, 8)), plate = 1) %>%
  mutate(plate_coord = paste0(plate, "_", plate_row, plate_col))

```

```{r upload filepaths}

epi_report_fp <- here("upload", "epi", paste(project_name, "PHL_sequencing_results.csv", sep = "_"))
seqsender_meta_fp <- here("upload", "seqsender", paste(project_name, "PHL2_seqsender_upload.csv", sep = "_"))
fasta_fp <- here("upload", "fasta", paste0(project_name, "_PHL2_combined.fasta"))

```

```{r save fastq and fasta for upload, eval = import_data}

#get only the good quality samples to upload
s_upload <- s_toPlot %>%
  filter4report() %>%
  select(sample_id) %>%
  pull()

if(length(s_upload) > 0) {

  #find all scrubbed fastq files
  dir.create(here("upload", "fastq"), recursive = TRUE)
  
  fastq_fps <- list.files(here("data", "processed_bs"), pattern = "scrubbed.*_001.fastq.gz", recursive = TRUE, full.names = TRUE)
  
  fastq2upload <- fastq_fps[grepl(paste0(s_upload, collapse = "|"), fastq_fps)]
  invisible(sapply(fastq2upload, function(x) file.rename(x, to = paste0(here("upload", "fastq"), gsub(".*scrubbed\\.[0-9]*\\.", "/", x)))))
  
  
  #find all fasta files
  dir.create(here("upload", "fasta"), recursive = TRUE)
  
  fasta_fps <- list.files(here("data", "processed_bs"), pattern = ".consensus_hard_masked_sequence.fa", recursive = TRUE, full.names = TRUE)
  
  fasta2upload <- fasta_fps[grepl(paste0(s_upload, collapse = "|"), fasta_fps)]

  fasta2upload %>%
    data_frame(FileName = .) %>%
    group_by(FileName) %>%
    do(read_delim(.$FileName, col_names = FALSE, delim = "fake_delim")) %>%
    ungroup() %>%
    select(-c(FileName)) %>%
    write_delim(file = fasta_fp, delim = "", col_names = FALSE)
  
}

```

```{r}
```
