---
title: |
  ![](forKnitting/logo_blk.png){width=5in}  
  "`r paste0('Sequencing run QC report')`"
author: "Philadelphia Public Health Labs"
date: \today
geometry: margin=3cm
output: 
    pdf_document:
        template: forKnitting/toc_after.tex
        keep_tex: false
        toc: true
        toc_depth: 3
        includes:
            in_header: forKnitting/TeX_packages_commands.sty

---

\tableofcontents

<!-- ========================================================== -->
<!--   Beginning of Preamble : Preamble seldom requires change  -->
<!-- ========================================================== -->

```{r setup, echo=FALSE}
### ================
###   knitr setup
### ================
library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=FALSE,
  cache.lazy = FALSE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  dpi=100,
  fig.width=6,
  fig.height=6,
  fig.align = "center",
  dev.args = list(pdf = list(useDingbats = FALSE))
  )

pander::panderOptions("knitr.auto.asis", FALSE)
#this also means that you have to put results='asis' in any block that has pander output
#this lets you loop through variables and produce multiple pander tables and ggplots in a
#single code block!

```

```{r child = 'Rmd/preamble.Rmd'}
```


```{r, Samples error check 1, eval=FALSE}
### ===========================
###   check for missing samples
### ===========================

### possible issue 1: Samples found in the sample sheet but not in the feature table (0 reads)
s_missing <- s %>%
  filter(!SampleID %in% colnames(cts)) %>%
  select(SampleID, sample_type, isControl)

if (any(!s_missing$isControl)) {
  pander(filter(s_missing, !isControl), caption="These samples were in the sample sheet but not in the feature table.")
  #stop (simpleError("Please fix"))
}

```


```{r, Samples error check 2, eval = FALSE}
### possible issue 2: Samples found in the feature table but not in the sample sheet. There must be an error!
in_counts_not_in_s <- setdiff(colnames(cts), s$SampleID)
if (length(in_counts_not_in_s) > 0) {
  stop (simpleError("These SampleID(s) are in the feature table, but not found in the sample sheet.", paste(in_counts_not_in_s, collapse=" ")))
}
```

\newpage

# Introduction

This report summarizes the results of `r nrow(s)` COVID samples sequenced on `r run_date` and analyzed with BaseSpace DRAGEN Lineage App version `r basespace_version` on `r samp_sh_header$Date`. Results section contain lineage and clade information using panogolin version `r pangolin_version` (https://cov-lineages.org/resources/pangolin.html) and Nextclade version `r nextclade_version` (https://clades.nextstrain.org/)

# FASTQ generation, demultiplexing, and quality control

## bcl2fastq

BaseSpace uses the bcl2fastq command to generate FASTQ files from BCL files (raw data files that contain sequence information and quality scores). bcl2fastq version `bcl2fq_version` was used to generate the FASTQ files using the command: \newline
`r bcl2fq_flags`

## Number of read pairs per sample after demultiplexing

Samples were sequenced on the `r sample_sh_header$instrument_type` as a `r paste0(unlist(sample_sheet$Reads), collapse = ":")` run and demultiplexed on BaseSpace by the bcl2fastq command. The demultiplexing step involves matching the barcode sequences associated with each sample to the sequence each read is tagged with

```{r percent of reads that passed filter, echo=FALSE}
s %>%
  mutate(pf_percent=total_pass_filter_reads/total_raw_reads) %>%
  ggplot(aes(x=pf_percent, fill = sample_type)) +
    geom_histogram(binwidth=0.1, boundary=TRUE, color = "white") +
    scale_fill_manual(values = ann_colors$sample_type) +
    scale_x_continuous(labels = scales::percent) +
    theme_bw() + 
    theme(panel.grid = element_blank()) +
    labs(
      x="Percent of reads that passed filter",
      y="Number of samples", fill = "Sample Type"
    )
```

```{r reads_histogram, echo=FALSE}
s %>%
  mutate(num_seq=total_pass_filter_reads/1000000) %>%
  # remove reads that did not get assigned
  filter(SampleID != "None") %>%
  ggplot(aes(x=num_seq, fill = sample_type)) +
    geom_histogram(binwidth=0.5, boundary=TRUE, color = "white") +
    scale_fill_manual(values = ann_colors$sample_type) +
    theme_bw() + 
    theme(panel.grid = element_blank()) +
    labs(
      x="Number of read pairs in sample (millions, M)",
      y="Number of samples",
      fill = "Sample Type"
    )
```

```{r percent reads of lane, echo=FALSE}
reads_percent <- s %>%
  mutate(percent_of_lane = total_pass_filter_reads/sum(total_pass_filter_reads)) %>%
  group_by(sample_type) %>%
  mutate(tot_percent = sum(percent_of_lane)) %>%
  ungroup() %>%
  order_on_other_col(sample_type, tot_percent, decreasing = FALSE) %>%
  order_on_other_col(SampleID, percent_of_lane) %>%
  arrange(SampleID)

reads_percent %>%
  ggplot(aes(x = "", y = percent_of_lane, fill=sample_type)) +
    geom_col(color = "white") +
    scale_y_continuous(labels = scales::percent) +
    scale_fill_manual(values = ann_colors$sample_type[match(levels(reads_percent$sample_type), names(ann_colors$sample_type))]) +
    theme_bw() +
    theme(panel.grid = element_blank(),
          axis.ticks.x.bottom = element_blank()) +
    labs(
      x="",
      y="Percent of assigned reads by sample type",
      fill = "Sample Type"
    )

```

## Top 10 unassigned indices

```{r}

top_unkwn_indices %>%
  ggplot(aes(y=fct_rev(index_sequence), x=number_of_reads)) +
  geom_col() +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  labs(
      x="Number of reads",
      y="Unassigned index sequences"
  )

```

These are the adapters: \newline
`r paste0(gsub(",", ": ", unlist(sample_sheet$Settings)), collapse = "\n")`

\blandscape

## Read counts and final library concentration of each sample

```{r, fig.width=10, fig.height=5}



s %>%
  mutate(Read_Counts = ifelse(is.na(nonhost), 0, nonhost)) %>%
  droplevels() %>%
  ggplot(aes(x = final_library_conc_ng_ul, y = Read_Counts, fill = sample_type)) +
    geom_point(shape = 21) +
    geom_hline(yintercept = min_reads, color="black", linetype="dashed") +
    theme_bw() + 
    scale_y_continuous(trans = "log10") +
    guides(fill = guide_legend(override.aes = list(shape = 21))) +
    labs(y = "Read Counts", x = "Final library DNA conc (ng/uL)")

```
\newpage
## Library concentration and read counts per plate

The samples are processed on a 96-well plate. The figures illustrate the library concentration, read counts after demultiplexing and read counts after quality control / denoising steps for each well. The gray color represents NA values. 

```{r}
s %>%
  mutate(plate = paste("Plate", plate)) %>%
  ggplot(aes(x=plate_column, y=fct_rev(plate_row), fill = final_library_conc_ng_ul)) +
    geom_tile(fill = "white") +
    geom_point(shape = 21, size = 10) +
    facet_wrap(~plate, ncol = 1) +
    scale_fill_viridis_c() +
    scale_x_continuous(breaks=seq(1,12), expand=c(0,0), limit = c(0, 13), position = "top") +
    theme_bw() +
    theme(
      strip.background = element_blank(),
      strip.placement = "outside",
      panel.grid = element_blank(),
      aspect.ratio = 0.66
    ) +
    labs(
      x="",
      y="Plate row",
      fill="Library\nconcentration\n(ng/ul)"
    )
```

```{r}
s %>%
  mutate(plate = paste("Plate", plate)) %>%
  ggplot(aes(x=plate_column, y=fct_rev(plate_row), fill = input)) +
    geom_tile(fill = "white") +
    geom_point(shape = 21, size = 10) +
    facet_wrap(~plate, ncol = 1) +
    scale_fill_viridis_c() +
    scale_x_continuous(breaks=seq(1,12), expand=c(0,0), limit = c(0, 13), position = "top") +
    theme_bw() +
    theme(
      strip.background = element_blank(),
      strip.placement = "outside",
      panel.grid = element_blank(),
      aspect.ratio = 0.66
    ) +
    labs(
      x="",
      y="Plate row",
      fill="Raw\nread\ncounts"
    )
```


```{r}
s %>%
  mutate(plate = paste("Plate", plate)) %>%
  ggplot(aes(x=plate_column, y=fct_rev(plate_row), fill = nonhost)) +
    geom_tile(fill = "white") +
    geom_point(shape = 21, size = 10) +
    facet_wrap(~plate, ncol = 1) +
    scale_fill_viridis_c() +
    scale_x_continuous(breaks=seq(1,12), expand=c(0,0), limit = c(0, 13), position = "top") +
    theme_bw() +
    theme(
      strip.background = element_blank(),
      strip.placement = "outside",
      panel.grid = element_blank(),
      aspect.ratio = 0.66
    ) +
    labs(
      x="",
      y="Plate row",
      fill="QC\nread\ncounts"
    )

do a detection for each plate

```

DRAGEN COVID Lineage App (on BaseSpace cloud) version `r dragen_covid_lineage_version` \newline
Panogolin database version `r panogolin_database_version` \newline
Lineage assignment software and version `r lineage_assignment_software_version`

\elandscape

# Appendix

## Number of reads before and after trimmming Illumina adapter sequences with Trimmomatic.

```{r trimmed reads, echo=FALSE}

preprocess %>%
  merge(select(s, SampleID, sample_type), by = "SampleID", all.x = TRUE) %>%
  arrange(-both_kept) %>%
  select(SampleID,
         sample_type,
         Input = input,
         Dropped = dropped,
         `Both kept` = both_kept) %>%
  kable_style()
```

\newpage

## Number of reads before and after filtering of host genome sequence.

```{r filtered reads, echo=FALSE}
preprocess %>%
  merge(select(s, SampleID, sample_type), by = "SampleID", all.x = TRUE) %>%
  mutate(`Percent host reads` = 100 * host / (host + nonhost),
    `Percent host reads` = round(`Percent host reads`, 2)) %>%
  arrange(`Percent host reads`) %>%
  select(SampleID,
         sample_type,
        `Host reads` = host,
        `Non-host reads` = nonhost,
        `Percent host reads`) %>%
  kable_style()
```


```{r Generate time stamped report, eval=FALSE, include=FALSE}
#notes
#neat: you can run this following command in the console to give your reports custom names (or date-stamp them)
#rmarkdown::render('ronen_stein_Report_shotgun.Rmd',output_file = paste('CEASE.report.', Sys.Date(), '.pdf', sep=''))
library(here)
rstudioapi::documentSaveAll()
dir.create(here("output", "basic_report"))

sequencing_run_date <- gsub("_.*", "", basename(here()))
project_name <- sub("^([^_]+)_", "", basename(here()), perl = TRUE)

qc_report_fp <- here(paste(sequencing_run_date, "_" , project_name, "_QC_Report.Rmd", sep=""))
qc_output_fp <- here("output", "basic_report", paste(project_name, '.QC.report.', Sys.Date(), '.pdf', sep=''))

basic_report_fp <- here("Rmd", "basic_report.Rmd")
basic_output_fp <- here("output", "basic_report", paste(project_name, '.basic.report.', Sys.Date(), '.pdf', sep=''))

rmarkdown::render(qc_report_fp, output_file = qc_output_fp, "all")
rmarkdown::render(basic_report_fp, output_file = basic_output_fp, "all")

```



