---
title: |
  ![](../forKnitting/logo_blk.png){width=5in}  
  "`r paste0('Basic report for ', gsub("_", " ", basename(here())))`"
author: "PennCHOP Microbiome Program"
date: \today
geometry: margin=3cm
output: 
    pdf_document:
        template: ../forKnitting/toc_after.tex
        keep_tex: false
        toc: true
        toc_depth: 3
        includes:
            in_header: ../forKnitting/TeX_packages_commands.sty

---

\tableofcontents

<!-- ========================================================== -->
<!--   Beginning of Preamble : Preamble seldom requires change  -->
<!-- ========================================================== -->

```{r setup, echo=FALSE}
### ================
###   knitr setup
### ================
library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=FALSE,
  cache.lazy = FALSE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  dpi=100,
  fig.width=7,
  fig.height=8,
  fig.align = "center",
  dev.args = list(pdf = list(useDingbats = FALSE))
  )

pander::panderOptions("knitr.auto.asis", FALSE)
#this also means that you have to put results='asis' in any block that has pander output
#this lets you loop through variables and produce multiple pander tables and ggplots in a
#single code block!

```

```{r child = 'preamble.Rmd'}
```

\newpage

# Introduction

This report is based on the results of sequencing performed on `r run_date` for `r investigator`

# Demultiplexing and quality control

## Number of read pairs per sample after demultiplexing

Samples were sequenced on Hiseq 2500 and demultiplexed. The demultiplexing step involves matching the barcode sequences associated with each sample to the sequence each read is tagged with.

```{r reads_histogram, echo=FALSE, fig.width=6, fig.height=6}
s %>%
  mutate(num_seq=input/1000000) %>%
  ggplot(aes(x=num_seq)) +
    geom_histogram(aes(fill=sample_type), binwidth=1, boundary=TRUE) +
    theme_bw() + 
    labs(
      x="Number of read pairs in sample (millions, M)",
      y="Number of samples"
    )
#ggsave(filename="read_qc.pdf", width=7, height=5, useDingbats=F)
```

\newpage

\blandscape

## Overall distribution of percentage reads removed in quality control

The low quality reads defined by Trimmomatic-0.33 were discarded from further analysis. Human DNA was filtered using BWA with HG38 version of human genome as reference. Reads mapping to the PhiX genome was also removed. Only the reads tagged as non-human were further analyzed.

```{r preprocess, echo=FALSE, fig.width=10, fig.height=5}
s %>%
  rbind(s) %>%
  mutate(low_quality = (input - host - nonhost) / input) %>%
  mutate(human = host / input) %>%
  mutate(non_human = nonhost / input) %>%
  ##can comment out this line after checking control quality as well
  #filter(!isControl) %>%
  arrange(desc(human)) %>%
  mutate(Sample_num=row_number()) %>%
  melt(c("Sample_num", quality_by), c("low_quality", "human", "non_human")) %>%
  ggplot(aes(x=Sample_num, y=value)) +
    geom_area(aes(fill=variable), position='stack') + 
    facet_grid(.~eval(parse(text=quality_by)), scales = "free_x") +
    scale_fill_brewer(palette="Set1") + 
    theme(axis.text.x = element_blank(), axis.ticks = element_blank()) +
    scale_x_continuous(expand=c(0,0)) +
    scale_y_continuous(expand=c(0,0), labels=scales:::percent) +
    labs(x="Samples", y="Percentage of reads", fill="")
#ggsave(filename='preprocess_summary.pdf', width=5, height=7, useDingbats=F)
```

\newpage
## Read counts and final library concentration of each sample

```{r, fig.width=10, fig.height=5}

s %>%
  mutate(Read_Counts = ifelse(is.na(nonhost), 0, nonhost)) %>%
  droplevels() %>%
  ggplot(aes(x = final_library_conc_ng_ul, y = Read_Counts, fill = sample_type)) +
    geom_point(shape = 21) +
    geom_hline(yintercept = min_reads, color="black", linetype="dashed") +
    theme_bw() + 
    scale_y_continuous(trans = "log10") +
    guides(fill = guide_legend(override.aes = list(shape = 21))) +
    labs(y = "Read Counts", x = "Final library DNA conc (ng/uL)")

```

\elandscape

# Taxonomic assignments

```{r heatmap_assignments}

satu_limit <- 0.4

```

Taxonomic assignments were performed using the Kraken program.

Heatmap charts were generated from the taxonomic assignments. Each column represents one sample and each row represents one taxon (typically a species). Ranks are included in the plot if the taxon is present in `r 100*prop_cut`% abundance in at least one sample.

The chart is colored white if species were not observed in the sample, dark blue if species were observed at very low abundance.  This allows the reader to quickly survey species presence/absence.  Abundance values exceeding `r 100*satu_limit`% are colored red, indicating an extremely dominant species.

Please note: only bacterial results are shown.\
`r if(show.text){paste0("Please see attached plot ", heatmap_file, " for the full heatmap")}`

```{r heatmap time, fig.height=10, fig.width=10}

# select taxa with mean relative abundance of prop_cut in at least one sample type
select_taxa <- summed_props %>%
  as.data.frame() %>% 
  rownames_to_column("Taxa") %>% 
  filter(!grepl("phage|sapiens", Taxa)) %>%
  pivot_longer(-Taxa, names_to="SampleID", values_to="props") %>%
  right_join(s_toPlot, by="SampleID")  %>%
  group_by(sample_type, Taxa) %>%
  mutate(mean_prop = mean(props)) %>%
  ungroup() %>%
  filter(mean_prop > prop_cut) %>%
  mutate(phyla = gsub(" .*", "", Taxa)) %>%
  mutate(phyla = factor(phyla, levels = unique(phyla))) %>%
  mutate(phyla = fct_relevel(phyla, "Actinobacteria",
                                    "Bacteroidetes",
                                    "Firmicutes",
                                    "Proteobacteria",
                                    "Verrucomicrobia")) %>%
  arrange(phyla) %>%
  pull(Taxa) %>%
  as.character() %>%
  unique()

if(!show.text) {

  summed_props[select_taxa, s_toPlot$SampleID] %>%
    pheat() %>%
    pheat_color_saturated(saturation_limit = satu_limit) %>%
    pheat_annotate_cols(s_toPlot[, c("SampleID", quality_by, color_by, shape_by)]) %>%
    pheat_display_cols(gaps = factor_gaps(paste0(s_toPlot[, quality_by], s_toPlot[, color_by])))
}

### top species
# top <- top_table(summed_props[select_taxa, s_toPlot$SampleID], s_toPlot, option=2, prop_cut = prop_cut)
# 
# top$avg_prop <- apply(top,1,mean)
# 
# top <- top %>%
#   rownames_to_column() %>%
#   rename(Taxa = rowname) %>%
#   filter(!row.names(top) %in% c("g__; s__")) %>%
#   mutate(Taxa = gsub("g__","",Taxa)) %>%
#   mutate(Taxa = gsub("; s__"," ", Taxa)) %>%
#   arrange(desc(avg_prop))

```

`r if(!has_pos_ctrl){paste0("\\textcolor{red}{\\huge THERE IS NO VIBRIO IN THE POSITIVE CONTROLS}")}`

\newpage

# Beta diversity

## Bray-Curtis distance

### PCoA plot based on Bray-Curtis distance

Here, we use Bray-Curtis distance to compare the species composition of the samples to each other.

The first plot shows the distance between each pair of samples in a single 2D plot.  It is not possible to plot the distances exactly on paper, so we have used a method of ordination called Principal Coordinates Analysis to select the best coordinate system for display.  The percentage of total variance captured along each axis is displayed on the chart.

```{r bray-curtis, fig.height=5, fig.width=7}

s_toPlot %<>%
  filter(Keep) %>%
  filter(!isControl)

dist_in <- usedist::dist_subset(bc, s_toPlot$SampleID)
plot(make_pcoa_plot(dm = dist_in, s = s_toPlot, color_by=color_by, shape_by=quality_by))

#manual
#plot(make_pcoa_plot(dist_in, s_toPlot, color_by="sample_type", shape_by="expected_genus"))
```

\newpage

## Jaccard distance

Here, we use Jaccard distance to compare samples based on shared species membership.  Plots are described above.

### PCoA plot based on Jaccard distance

```{r jaccard, fig.height=5, fig.width=7}

dist_in <- usedist::dist_subset(jd, s_toPlot$SampleID)
plot(make_pcoa_plot(dist_in, s_toPlot,  color_by=color_by, shape_by=quality_by))
```

\newpage

# Alpha diversity

Here, we show richness (# of species at a rarefying level of 1,000 reads) and Shannon diversity index (number of species weighted by abundance) between groups. \newline
Higher shannon diversity means greater "eveness" (e.g. one hugely abundant species and 100 scarce ones has a lower shannon value than 101 equally abundant species) see this [*page*](https://junglee0713.netlify.com/2018/12/12/shannon-diversity-index/) for a great interactive explanation.


```{r alpha diversity, fig.height=8, fig.width=6}

s_toPlot %>%
  filter(Keep) %>%
  filter(!isControl) %>%
  get_alpha(cts) %>%
  ggplot(aes(x=eval(parse(text=color_by)), y=alpha, color=eval(parse(text=color_by)))) +
  geom_boxplot() +
  facet_grid(metric~eval(parse(text=quality_by)), scales = "free") +
  labs(y="", x=color_by, color=color_by) +
  theme_bw() +
  theme(axis.text.x=element_text(angle=-25, hjust= .1)) +
  #guides(color=F) +
  scale_color_brewer(palette = "Set2")
```

\newpage

# Appendix

## Number of reads before and after trimmming Illumina adapter sequences with Trimmomatic.

```{r trimmed reads, echo=FALSE}

preprocess %>%
  merge(select(s, SampleID, sample_type), by = "SampleID", all.x = TRUE) %>%
  arrange(-both_kept) %>%
  select(SampleID,
         sample_type,
         Input = input,
         Dropped = dropped,
         `Both kept` = both_kept) %>%
  kable_style()
```

\newpage

## Number of reads before and after filtering of host genome sequence.

```{r filtered reads, echo=FALSE}
preprocess %>%
  merge(select(s, SampleID, sample_type), by = "SampleID", all.x = TRUE) %>%
  mutate(`Percent host reads` = 100 * host / (host + nonhost),
    `Percent host reads` = round(`Percent host reads`, 2)) %>%
  arrange(`Percent host reads`) %>%
  select(SampleID,
         sample_type,
        `Host reads` = host,
        `Non-host reads` = nonhost,
        `Percent host reads`) %>%
  kable_style()
```
