---
title: |
  ![](../forKnitting/logo_blk.png){width=5in}  
  "`r paste0('Analytical report for ', gsub("_", " ", basename(here())))`"
author: "PennCHOP Microbiome Program"
date: \today
geometry: margin=3cm
output: 
    pdf_document:
        template: ../forKnitting/toc_after.tex
        keep_tex: false
        toc: true
        toc_depth: 3
        includes:
            in_header: ../forKnitting/TeX_packages_commands.sty

---

\tableofcontents

<!-- This is a mother report and should be the only report that has a Yaml and knitr setup block -->
<!-- See https://martinctc.github.io/blog/first-world-problems-very-long-rmarkdown-documents/ -->

<!-- ========================================================== -->
<!--   Beginning of Preamble : Preamble seldom requires change  -->
<!-- ========================================================== -->

```{r setup, echo=FALSE}
### ================
###   knitr setup
### ================
library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=FALSE,
  cache.lazy = FALSE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  dpi=100,
  fig.width=8,
  fig.height=8,
  fig.align = "center",
  dev=c("png", "pdf"),
  dev.args = list(pdf = list(useDingbats = FALSE))
  )

pander::panderOptions("knitr.auto.asis", FALSE)
#this also means that you have to put results='asis' in any block that has pander output
#this lets you loop through variables and produce multiple pander tables and ggplots in a
#single code block!

```

```{r child = 'preamble.Rmd'}
```

# Introduction

This report is based on the results of sequencing performed on `r run_date` for `r investigator`


```{r}
heatmap_fp <- here("output", Sys.Date(), paste(Sys.Date(), "_taxonomy_heatmap.pdf", sep=''))
heatmap_filename <- paste(Sys.Date(), "_taxonomy_heatmap.pdf", sep='')
dir.create(here("output", Sys.Date()))
show.text <- nrow(s) > 40
satu_limit <- 0.4
```
\newpage
# Taxonomic assignments

Taxonomic assignments were performed using the Kraken program.

Heatmap charts were generated from the taxonomic assignments. Each column represents one sample and each row represents one taxon (typically a species). Ranks are included in the plot if the taxon is present above `r 100*prop_cut`% mean abundance across all samples.

The chart is colored white if species were not observed in the sample, dark blue if species were observed at very low abundance.  This allows the reader to quickly survey species presence/absence.  Abundance values exceeding `r 100*satu_limit`% are colored red, indicating an extremely dominant species.

Please note: only bacterial results are shown.

`r if(show.text){paste0("Please see attached plot ", heatmap_filename, ".")}`

```{r heatmap time, fig.height=10, fig.width=10}

annotations <- s_toTest %>%
  select(SampleID, sample_type, sg_toTest, study_day) %>%
  arrange(eval(parse(text=sg_toTest)))

select_taxa <- summed_props %>%
  as.data.frame() %>% 
  rownames_to_column("Taxa") %>% 
  pivot_longer(-Taxa, names_to="SampleID", values_to="props") %>%
  right_join(s_toTest, by="SampleID")  %>%
  group_by(Taxa) %>%
  mutate(mean_props = mean(props)) %>%
  ungroup() %>%
  filter(mean_props > prop_cut) %>%
  mutate(phyla = gsub(" .*", "", Taxa)) %>%
  mutate(phyla = factor(phyla, levels = unique(phyla))) %>%
  mutate(phyla = fct_relevel(phyla, "Actinobacteria",
                                    "Bacteroidetes",
                                    "Firmicutes",
                                    "Proteobacteria",
                                    "Verrucomicrobia")) %>%
  arrange(phyla) %>%
  pull(Taxa) %>%
  as.character() %>%
  unique()

heatmap_props <- summed_props[select_taxa, annotations$SampleID]

heatmap_props %>%
    pheat() %>%
    pheat_color_saturated() %>%
    pheat_annotate_cols(annotations) %>%
    pheat_display_cols(gaps = factor_gaps(annotations[, sg_toTest])) %>%
    #using if statement here because ifelse is vectorized
    #have to include brackets around if statement otherwise if will always be true (1 is passed)
    {`if`(show.text, pheat_save(., heatmap_fp), .)}

#include pheat_annotation_color to color pheatmap when doing analysis

### top species
# top <- top_table(heatmap_props, s_toTest, option=2, prop_cut = prop_cut)
# 
# top$avg_prop <- apply(top,1,mean)
# 
# top <- top %>%
#   rownames_to_column() %>%
#   rename(Taxa = rowname) %>%
#   filter(!row.names(top) %in% c("g__; s__")) %>%
#   mutate(Taxa = gsub("g__","",Taxa)) %>%
#   mutate(Taxa = gsub("; s__"," ", Taxa)) %>%
#   arrange(desc(avg_prop))

```

\newpage
`r paste0(sg_toTest)` to test

```{r study group samples}

table(s_toTest$sample_type, s_toTest[, sg_toTest]) %>%
   kable("latex", longtable = F, digits=2, booktabs=T, escape=F) %>%
   kable_styling(full_width = F, position = "left") %>%
   row_spec(0, bold = T, color="#7C0A02")

```

\newpage
# Relative abundances

```{r differential abundances}

for(g in sg_toTest) {
  p <- props_toTest %>%
    mutate(Taxa = sub(" ", "\n", Taxa)) %>%
    ggplot(aes(x=eval(parse(text=g)), y=props, fill=eval(parse(text=g)), shape = study_day)) +
      geom_boxplot(outlier.alpha=0) +
      geom_point(size = 0.5, position = position_jitterdodge()) +
      facet_wrap(~Taxa, scales="free_y") +
      scale_fill_manual(values=ann_colors[[g]]) +
      #scale_shape_manual(values=c(15, 16, 17, 18, 8)) +
      scale_y_continuous(labels=scales:::percent) +
      theme_bw() +
      theme(
        panel.grid = element_blank(),
        strip.background = element_blank(),
        axis.text.x=element_text(angle=-25, hjust= .1)
      ) +
      guides(color=F) +
      labs(
        x="", shape = "Study day",
        y="Relative abundance",
        fill = cap_first(g)
      )
  
  print(p)
  cat("\n\n")
  
}

```

```{r, eval = FALSE}

#use jtools to plot estimate and variance of models and see which random effect terms to use, based off of shannon
#only include terms in random effects that explain a lot of the variance

s_toTest_fit <- s_toTest %>%
  get_alpha(cts) %>%
  pivot_wider(names_from = "metric", values_from = "alpha")

#compare models by changing one random effect at a time
fit <- lme4::lmer(Shannon ~ study_group + (1|subject_id), data = s_toTest_fit, na.action = na.omit)
fit2 <- lme4::lmer(Shannon ~ study_group + (1|subject_id) + (1|study_day/subject_id), data = s_toTest_fit, na.action = na.omit)
#can also use summ to look at how much std dev is assigned to each random variable. Want the least std dev to go to residual
summ(fit2)
plot_summs(fit, fit2, scale = T)

#use cage_id and fmt_expt as random effects since it explains most of the variance compard to other random effects. Don't use donors because it explains basically 0 variance when cage_id is included. Donors explains a lot of the vaiance only with Group term is used, but thats expected and they are essentially the same groups
```

\newpage
Linear model of Taxa and `r paste0(sg_toTest)`. Results are filtered by p.value < 0.05

```{r diff abund lme}

summaries2print <- data.frame(Value = double(), Std.Error = double(), 
                              DF = double(), t.value = double(), p.value = double(), Analysis = factor())

random_list <- list(~1|subject_id)

for(g in sg_toTest) {
  
  lme_form <- as.formula(paste0("props_log~", g))
  
  first_level <- levels(as.data.frame(props_toTest)[, g])[1]
  
  summaries_df <- props_toTest %>%
    group_by(Taxa) %>%
    #do(tidy(lm(lme_form, data=.))) %>%
    do(tidy_lmer(nlme::lme(lme_form, data=., random=random_list, na.action=na.omit))) %>%
    ungroup() %>%
    filter(!grepl("Intercept", term)) %>%
    get_fdr(term) %>%
    mutate(term = gsub(g, paste0(first_level, " - "), term)) #%>% #replace term with the first level of factor
    #filter(p.value<0.05)

  summaries2print <- rbind(summaries2print, summaries_df)


}
summaries2print %>%
  kable_sort(term) %>%
  kable_style()


```

\newpage
Post hoc test for pairwise comparisons of `r paste0(sg_toTest)` on Taxa

```{r diff abund lme posthoc, eval=FALSE}

summaries2print <- data.frame(Value = double(), Std.Error = double(), 
                              DF = double(), t.value = double(), p.value = double(), Analysis = factor())

for(g in sg_toTest) {
  
  summaries_df <- props_toTest %>%
    group_by(Taxa) %>%
    #have to put the actual formula in this line for emmeans to work
    do(tidy_lmer2(nlme::lme(as.formula(paste0("props_log~", g)), random=random_list, na.action=na.omit, data=.), g)) %>%
    ungroup() %>%
    #mutate(Analysis = g) %>%
    filter(!grepl("Intercept", contrast)) %>%
    filter(!grepl(g, contrast)) %>%
    select(-c(numDF, denDF, F.value, contrast)) %>%
    filter(!is.na(estimate))  %>%
    rename(contrast = "X1") %>% ##emmeans is giving X1 as column name for comparisons
    get_fdr(contrast)

  summaries2print <- rbind(summaries2print, summaries_df)
}

summaries2print %>%
  select(Taxa, contrast, estimate, df, SE, t.ratio, p.value, fdr) %>%
  kable_sort(contrast) %>%
  kable_style()
  
```

\newpage

# Beta diversity

## Bray-Curtis distance

PCoA plot based on Bray-Curtis distance. Ellipses were drawn with a t-distribution.


```{r bray_curtis pcoa, fig.height=4, fig.width=6}

for(g in sg_toTest) {

  dist_toTest <- dist_subset(bc, s_toTest$SampleID)
  pc <- pcoa(dist_toTest)
  pc_df <- merge(s_toTest, pc$vectors[, 1:3], by.x="SampleID", by.y="row.names")
  pct <- round(pc$values$Relative_eig * 100)

  p <- pc_df %>%
    ggplot(aes(x=Axis.1, y=Axis.2, color=eval(parse(text=g)))) +
      geom_point(aes(shape = study_day), size=2) +
      stat_ellipse() +
      scale_color_manual(values=ann_colors[[g]]) +
      #scale_shape_manual(values=c(15, 16, 17)) + 
      theme_bw() +
      theme(
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid = element_blank(), 
        aspect.ratio = 1,
        strip.background = element_blank()
      ) + 
      labs(x=paste0("PCoA axis 1 (", pct[1], "%)"), 
           y=paste0("PCoA axis 2 (", pct[2], "%)"),
           color=cap_first(g), shape="Study day", lty ="", title="Bray Curtis")

  print(p)
  cat("\n\n")
}

```

Bray-Curtis distance PERMANOVA analysis

```{r bc permanova}

#just use permanova first with random terms to see which random term explains most of the variance. Order of the terms matter so if one term explains most of the variance, the subsequent terms will not explain much variance and will be ignored by adonis (if this is the case, just use the main term). Keep terms that make residual explain the least variance (R2). 
#p_test <- pc_df %>% do(tidy_permanova(adonis(as.formula(paste0("dist_toTest ~ ", g, "+ cage_id + subject_id")), data=., permutations=99)))

#if want to use shuffle, first calculate if have enough cages for permutations by doing choose(# of cages, # of groups to test). If this number is below 20, just group by the desired testing groups and use adonis with random variable to do analysis
#also, permutation should be below this number so it doesn't count the randomly shuffled groups twice

##define a dataframe so posthoc tests can be added as new rows
p_anova <- data.frame(Group = factor(),
                      Term = factor(),
                      Df = numeric(),
                      SumsOfSqs = numeric(),
                      MeanSqs = numeric(),
                      F.Model = numeric(),
                      R2 = numeric(),
                      p.value=numeric())

covars <- gsub("~1 \\| ", "", paste0(unlist(random_list)))[2:length(random_list)]

for(g in sg_toTest) {
  
  summaries_df <- dist_toTest %>%
    permanova_with_shuffle_1_group_posthoc(., s_toTest, g, 'subject_id', covars,  99, is_within = F) %>%
    filter(!is.na(p.value))# %>%
    #mutate(Group = g)
  
  ##if not enough permutations
  #s_toTest %>%
     #do(tidy_permanova(adonis(as.formula(paste0("dist_toTest ~ ", g, " + subject_id")), data=., permutations=99)))
  
  p_anova <- rbind(p_anova, summaries_df)
  
}

p_anova %>%
  kable_sort() %>%
  kable_style()

```

\newpage

## Jaccard distance

Here, we use Jaccard distance to compare samples based on shared species membership.  Plots are described above.

PCoA plot based on Jaccard distance

```{r jaccard pcoa, fig.height=4, fig.width=6}

for(g in sg_toTest) {

  dist_toTest <- dist_subset(jd, s_toTest$SampleID)
  pc <- pcoa(dist_toTest)
  pc_df <- merge(s_toTest, pc$vectors[, 1:3], by.x="SampleID", by.y="row.names")
  pct <- round(pc$values$Relative_eig * 100)

  p <- pc_df %>%
    ggplot(aes(x=Axis.1, y=Axis.2, color=eval(parse(text=g)))) +
      geom_point(aes(shape = study_day), size=2) +
      stat_ellipse() +
      scale_color_manual(values=ann_colors[[g]]) +
      #scale_shape_manual(values=c(15, 16, 17)) + 
      theme_bw() +
      theme(
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid = element_blank(), 
        aspect.ratio = 1,
        strip.background = element_blank()
      ) + 
      labs(x=paste0("PCoA axis 1 (", pct[1], "%)"), 
           y=paste0("PCoA axis 2 (", pct[2], "%)"),
           color=cap_first(g), shape="Study day", lty ="", title="Jaccard")

  print(p)
  cat("\n\n")
}

```

Jaccard distance PERMANOVA analysis

```{r jd permanova}

##define a dataframe so posthoc tests can be added as new rows
p_anova <- data.frame(Group = factor(),
                      Term = factor(),
                      Df = numeric(),
                      SumsOfSqs = numeric(),
                      MeanSqs = numeric(),
                      F.Model = numeric(),
                      R2 = numeric(),
                      p.value=numeric())

for(g in sg_toTest) {
  
  summaries_df <- dist_toTest %>%
    permanova_with_shuffle_1_group_posthoc(., s_toTest, g, 'subject_id', covars, is_within = F, 99) %>%
    filter(!is.na(p.value))# %>%
    #mutate(Group = g)
  
  ##if not enough permutations
  #s_toTest %>%
     #do(tidy_permanova(adonis(as.formula(paste0("dist_toTest ~ ", g, " + subject_id")), data=., permutations=99)))
  
  p_anova <- rbind(p_anova, summaries_df)
  
}

p_anova %>%
  kable_sort() %>%
  kable_style()

```

\newpage

# Alpha diversity

Here, we show richness (# of species at a rarefying level of 1,000 reads) and Shannon diversity index between groups 

Shannon diversity index is weighted by abundance where higher numbers mean greater "eveness" (e.g. one hugely abundant species and 100 scarce ones has a lower shannon value than 101 equally abundant species).

```{r alpha diversity, fig.height=4, fig.width=6}

for(g in sg_toTest) {

  p <- s_toTest %>%
    get_alpha(cts) %>%
    ggplot(aes(x=eval(parse(text=g)), y=alpha, fill=eval(parse(text=g)), shape = study_day)) +
      geom_boxplot(outlier.alpha = 0) +
      geom_point(size = 2, position=position_jitterdodge()) +
      scale_fill_manual(values = ann_colors[[g]]) +
      #scale_shape_manual(values=c(15, 16, 17)) +
      facet_wrap(~metric, scales = "free_y") +
      theme_bw() +
      theme(
        panel.grid = element_blank(),
        strip.background = element_blank()
      ) +
      labs(
        x=cap_first(g), y="Alpha diversity",
        shape = "Study day", fill = cap_first(g)
      ) +
    theme(axis.text.x=element_text(angle=-25, hjust= .1))

  print(p)
  cat("\n\n")

}

```

Here, we test the significant of `r paste0(sg_toTest)` on alpha diversity
```{r alpha diversity test}

summaries2print <- data.frame(Value = double(), Std.Error = double(), 
                              DF = double(), t.value = double(), p.value = double(), Analysis = factor())

for(g in sg_toTest) {
  lme_form <- as.formula(paste0("alpha~", g))
  
  summaries_df <- s_toTest %>%
    get_alpha(cts) %>%
    group_by(metric) %>%
    #do(tidy(lm(lme_form, data=.))) %>% ##for models with only independent variables
    do(tidy_lmer(nlme::lme(lme_form, random=random_list, data=.))) %>% 
    ungroup() %>% 
    filter(!grepl("Intercept", term)) %>%
    mutate(term = gsub(g, paste0(levels(s_toTest[, g])[1], " - "), term)) ##replace term with the first level of factor
  
  summaries2print <- rbind(summaries2print, summaries_df)
  
}

#if lm not using factor names, restart RStudio
summaries2print %>%
  kable_sort() %>%
  kable_style()
```

Here, we perform pairwise comparisons of the `r paste0(sg_toTest)` on alpha diversity

```{r alpha diversity test pair, eval = FALSE}

summaries2print <- data.frame(Value = double(), Std.Error = double(), 
                              DF = double(), t.value = double(), p.value = double(), Analysis = factor())

for(g in sg_toTest) {
  
  summaries_df <- s_toTest %>%
    get_alpha(cts) %>%
    group_by(metric) %>%
    #have to put formula in same line as lme to emmeans to get it
    do(tidy_lmer2(nlme::lme(as.formula(paste0("alpha~", g)), random=random_list, data=.), g)) %>%
    ungroup() %>%
    #mutate(Analysis = g) %>%
    filter(!grepl("Intercept", contrast)) %>%
    filter(!grepl(g, contrast)) %>%
    select(-c(numDF, denDF, F.value, contrast)) %>%
    filter(!is.na(estimate))  %>%
    rename(contrast = "X1") ##emmeans is giving X1 as column name for comparisons

  summaries2print <- rbind(summaries2print, summaries_df)

}

summaries2print %>%
  select(metric, contrast, estimate, SE, df, t.ratio, p.value) %>%
  kable_sort() %>%
  kable_style()

```

```{r Generate time stamped report, eval=FALSE, include=FALSE}
#notes
#neat: you can run this following command in the console to give your reports custom names (or date-stamp them)
#rmarkdown::render('ronen_stein_Report_shotgun.Rmd',output_file = paste('CEASE.report.', Sys.Date(), '.pdf', sep=''))

library(here)
rstudioapi::documentSaveAll()

dir.create(here("output", Sys.Date()))
sequencing_run_date <- gsub("_.*", "", basename(here()))
project_name <- sub("^([^_]+)_", "", basename(here()), perl = TRUE)

report_fp <- here("Rmd", "mother_shotgun_report.Rmd")
output_fp <- here("output",Sys.Date(), paste(project_name, '.report.', Sys.Date(), '.pdf', sep=''))

rmarkdown::render(report_fp, output_file = output_fp, "all")

```

