
```{r libraries, message=FALSE, warning=FALSE}
library(here)
library(tidyverse)
library(kableExtra)
library(ghibli)

```

```{r functions}

source(here("functions", "R_all_functions_v3.R"))

```

```{r versions and dates}

#if these numbers are missing, throw an error

pangolin_version <-

nextclade_version <- 
  
basespace_version <-
  
date_pipeline_was_run <- 
  
sequencing_machine <- "MiSeq"

```


```{r file paths}

### mapping file path
#try to get it automagically
mapping_file_fp <- here("metadata", list.files(here("metadata"), pattern = ".tsv|.csv"))

### bcl2fastq info
b2fq_c_fp <- here("Data", "bcl2fastq", "commands.tsv")
b2fq_v_fp <- here("Data", "bcl2fastq", "bcl2fastqVersion.stderr.txt")

### demultiplexing results from BaseSpace
adaptertrimming_fp <- here("Data", "demux_reads", "AdapterTrimming.txt")
demuxsum_fp <- here("Data", "demux_reads", "DemuxSummaryF1L1.txt")
fastqsum_fp <- here("Data", "demux_reads", "FastqSummaryF1L1.txt")
SampleSheet_fp <- here("Data", "demux_reads", "SampleSheetused.csv")

### kmer fp
kmer_fp <- here("Data", "kmer", "Data_export.csv")

### NextClade results fp
nextclade_fp = here("Data", "NextClade", "Data_export.csv")

### Pangolin
pangolin_fp <- here("Data", "Pangolin", "Data_export.csv")

```

```{r load bcl2fq}

bcl2fq_version <- read_delim(b2fq_v_fp, skip = 1, col_names = FALSE) %>%
  slice(1) %>%
  select(X2) %>%
  pull()

#just reading in command file as one column
bcl2fq_flags <- read_delim(b2fq_c_fp, col_names = FALSE, col_types = cols(.default = "c"), delim = ";") %>%
  slice(1) %>%
  mutate(X1 = gsub("^.*bcl2fastq", "bcl2fastq", X1)) %>%
  mutate(X1 = gsub(" --", "\n\t--", X1)) %>%
  mutate(X1 = gsub(" -", "\n\t-", X1)) %>%
  pull()

```

```{r load read numbers}

trimmin <- read_delim(adaptertrimming_fp) %>%
  filter(!is.na(TrimmedBases))

# read counts by sample number
fastq_sum <- read_delim(fastqsum_fp) %>%
  group_by(SampleNumber) %>%
  summarise(total_raw_reads = sum(NumberOfReadsRaw), total_pass_filter_reads = sum(NumberOfReadsPF)) %>%
  ungroup()

# SampleSheet.csv loading

sample_sheet <- load_sample_sheet(SampleSheet_fp)

samp_sh_header <- data.frame(X1 = unlist(sample_sheet$Header)) %>%
  mutate(col_names = gsub(",.*", "", X1)) %>%
  mutate(col_names = gsub(" ", "_", col_names)) %>%
  mutate(X1 = gsub(".*,", "", X1)) %>%
  pivot_wider(names_from = "col_names", values_from = "X1")

samp_sh_s <- data.frame(X1 = unlist(sample_sheet$Data), stringsAsFactors = FALSE) %>%
  #split each line into columns using row 1 as the column names
  separate(col = "X1", strsplit(as.character(slice(., 1)), ",")[[1]], sep = ",") %>%
  slice(-1) %>%
  rename(SampleName = "Sample_Name", SampleNumber = "BASESPACE_ONLY_ORIGINAL_SAMPLE_NUMBER") %>%
  mutate(Barcode_sequence = paste0(I7_Index_ID, "+", I5_Index_ID)) %>%
  select(SampleName, SampleNumber, I7_Index_ID, I5_Index_ID, Barcode_sequence, Sample_Type, Index_ID)

#merge sample names with read counts
samp_names <- read_delim(demuxsum_fp, col_names = FALSE) %>%
  filter(grepl("Sample", X1)) %>%
  t() %>%
  as.data.frame(stringsAsFactors = FALSE) %>%
  purrr::set_names(as.character(slice(., 1))) %>%
  slice(-1) %>%
  merge(fastq_sum, by = "SampleNumber", all = TRUE) %>%
  merge(samp_sh_s, by = c("SampleName", "SampleNumber"), all = TRUE) %>%
  mutate(percent_reads_passed_filter = total_pass_filter_reads/total_raw_reads) %>%
  mutate(percent_of_lane = total_pass_filter_reads/sum(total_pass_filter_reads))

top_unkwn_indices <- read_delim(demuxsum_fp, col_names = FALSE) %>%
  filter(is.na(X3)) %>%
  select(X1, X2) %>%
  filter(!is.na(X2)) %>%
  rename(index_sequence = "X1", number_of_reads = "X2") %>%
  mutate(number_of_reads = as.numeric(number_of_reads)) %>%
  order_on_other_col(index_sequence, number_of_reads) %>%
  slice(1:10) %>%
  droplevels()

#no exhaustive barcode sequence info and no FASTQC info? and no host read mapping info?

```

```{r load data}

kmer <- read_csv(kmer_fp)

n_clade <- read_csv(nextclade_fp)

pango <- read_csv(pangolin_fp)


#TODO: Add a check to see if positive control shows up as the correct variant


#look for Vibrio campbellii taxa as positive control
#has_pos_ctrl <- any(grepl("Vibrio campbellii", a))


```

```{r sample_sheet_import and columns, echo=FALSE}
lookup <- c(subject_id = "SubjectID",
            host_species = "HostSpecies",
            sample_type = "SampleType",
            final_library_conc_ng_ul = "final_library_concentration_ng_ul")
ctrl_regex <- "^EBneg.*|^Extract.*|^Vibriolambda.*|^Blank.*|^MockDNA.*|^DNAfreewater.*|^Geneblock.*|^Emptywell"

s <- read_csv(mapping_file_fp) %>%
  filter(rowSums(is.na(.)|.=="") != ncol(.)) %>%
  mutate(SampleID = as.character(SampleID)) %>%
  rename(any_of(lookup)) %>%
  filter(!grepl("#", SampleID)) %>%
  left_join(preprocess, by="SampleID") %>%
  mutate(isControl = grepl(ctrl_regex, SampleID, ignore.case = TRUE)) %>%
  mutate(Keep = input > min_reads) %>%
  mutate(Keep = ifelse(!SampleID %in% colnames(cts), FALSE, Keep)) %>%
  mutate(Keep = ifelse(is.na(input), FALSE, Keep)) %>%
  mutate(sample_type = levels(sample_type)[sample_type]) %>%
  #rearrange the sampletypes by ctrls first
  mutate(sample_type = factor(sample_type, levels =
                               unique(sample_type)[order(match(unique(sample_type), unique(sample_type)[grepl(ctrl_regex, gsub("-| ", "", unique(sample_type)), ignore.case = TRUE)]))])) %>%
  #arrange(as.numeric(levels(study_day))[study_day]) %>%
  droplevels()

color_by <- NULL
shape_by <- NULL
quality_by <- "sample_type"
potential_headers <- c("study_group", "study_day", "subject_id",
                       "current_antibiotics", "host_species", "cage_number") #pick 2
header_idx <- which(is.element(potential_headers, colnames(s)))

if(length(header_idx)>0){
  color_by <- potential_headers[header_idx[1]]
}
if(length(header_idx)>1){
  shape_by <- potential_headers[header_idx[2]]
}

all_dates <- gsub("-", "", as.character(unique(s$run_start_date)))[!is.na(unique(s$run_start_date))]
run_date <- paste(lapply(all_dates, change_date_format), collapse=', ')
investigator <- paste(levels(s$investigator), collapse = ", ")
investigator <- gsub("(^|[[:space:]])([[:alpha:]])", "\\1\\U\\2", investigator, perl=TRUE)

show.text <- nrow(s) > 40
```


```{r for basic report}
### ====================================
### Select specific groups to plot for QC
### ====================================
s_toPlot <- s %>%
  filter(input > 0 & !is.na(input)) %>%
  select(SampleID, quality_by, color_by, shape_by, final_library_conc_ng_ul, Keep, isControl) %>%
  arrange(eval(parse(text=quality_by)), eval(parse(text=color_by)), eval(parse(text=shape_by))) %>%
  mutate(SampleID = as.character(SampleID))

```

```{r prop to test}
### ====================================
### Select specific groups to test and color them
### ====================================

s_toTest <- s %>%
  filter(Keep) %>%
  filter(!isControl) %>%
  ##can use mutate and factor the levels to reorder groups
  mutate(study_group = factor(study_group, levels=unique(study_group))) %>%     
  ##changing order of study_groups
  mutate(SampleID = as.character(SampleID)) %>%
  mutate(subject_id = as.character(subject_id)) %>%
  mutate(study_day = factor(study_day, levels = sort(unique(study_day)))) %>%
  arrange(eval(parse(text=quality_by)), eval(parse(text=color_by)), eval(parse(text=shape_by))) %>%
  droplevels()

props_toTest <- summed_props %>%
  melt() %>%
  setNames(c("Taxa", "SampleID", "props")) %>%
  filter(SampleID %in% s_toTest$SampleID) %>%
  merge(select(s_toTest, SampleID, subject_id, cage_id, nonhost, quality_by, color_by, shape_by), by="SampleID", all.y = TRUE) %>%
  group_by(Taxa) %>%
  mutate(mean_props = mean(props)) %>%
  ungroup() %>%
  mutate(phyla = gsub(" .*", "", Taxa)) %>%
  mutate(phyla = factor(phyla, levels = unique(phyla))) %>%
  mutate(phyla = fct_relevel(phyla, "Actinobacteria",
                                    "Bacteroidetes",
                                    "Firmicutes",
                                    "Proteobacteria",
                                    "Verrucomicrobia")) %>%
  arrange(phyla) %>%
  filter(mean_props > prop_cut)

props_toTest %<>%
  mutate(props = props + min(filter(props_toTest, props>0)$props) / 10) %>%
  mutate(props_log = log10(props))

#set colors for the groups
project_samples <- unique(s_toTest$sample_type)
  
ann_colors <- list(
  sample_type = setNames(brewer.pal(name = "Pastel1", n = length(project_samples)), project_samples)
)

#variables to test and colors
sg_toTest <- c("study_group")

MononokeStart <- 3

for(g in sg_toTest) {
  lvls <- levels(s_toTest[, g])
  g_colors <- setNames(ghibli_palette("MononokeLight")[MononokeStart:(MononokeStart-1+length(lvls))], lvls)
  ann_colors[[g]] <- g_colors
}

#scales::show_col(ann_colors[[sg_toTest]])

```


```{r load pcoa}

bc <- vegdist(t(summed_props))
jd <- vegdist(t(summed_props), binary = T)

```

```{r}
```
