
```{r libraries, message=FALSE, warning=FALSE}
library(here)
library(tidyverse)
library(pander)
library(kableExtra)
library(ghibli)
library(RColorBrewer)

```

```{r functions}

source(here("functions", "R_all_functions_v3.R"))

```

```{r versions and dates manual inputs}

#if these variables are missing, throw an error

# YY-MM-DD
date_pipeline_was_run <- "23-12-31"

# N.N.N
dragen_covid_lineage_version <- "1.1.1"

# N.N.N
nextclade_version <- "0.0.0"

primer_select <- 3 # set variable as 1 for V3, 2 for V4, or 3 for V4.1

artic_primer_scheme <- c("V3", "V4", "V4.1")[primer_select]

bs_bed_file_used <- c("nCov-2019.bed", "artic_V4_primer_scheme_NC_045512.2.bed", "artic_v4.1_primers_nc045512.2.bed")[primer_select]

# minimum acceptable number of reads
min_reads <- 10000

run_number <- 1
```

```{r file paths}

### mapping file path
#try to get it automagically
mapping_file_fp <- here("metadata", list.files(here("metadata"), pattern = ".tsv|.csv"))

### bcl2fastq info
b2fq_c_fp <- here("Data", "bcl2fastq", "commands.tsv")
b2fq_v_fp <- here("Data", "bcl2fastq", "bcl2fastqVersion.stderr.txt")

### demultiplexing results from BaseSpace
adaptertrimming_fp <- here("Data", "demux_reads", "AdapterTrimming.txt")
demuxsum_fp <- here("Data", "demux_reads", "DemuxSummaryF1L1.txt")
fastqsum_fp <- here("Data", "demux_reads", "FastqSummaryF1L1.txt")
SampleSheet_fp <- here("Data", "demux_reads", "SampleSheetused.csv")

### kmer fp
kmer_fp <- here("Data", "kmer", "Data_export.csv")

### NextClade results fp
nextclade_fp = here("Data", "NextClade", "Data_export.csv")

### Pangolin
pangolin_fp <- here("Data", "Pangolin", "Data_export.csv")

```

```{r load bcl2fq}

bcl2fq_version <- read_delim(b2fq_v_fp, skip = 1, col_names = FALSE) %>%
  slice(1) %>%
  select(X2) %>%
  pull()

#just reading in command file as one column
bcl2fq_flags <- read_delim(b2fq_c_fp, col_names = FALSE, col_types = cols(.default = "c"), delim = ";") %>%
  slice(1) %>%
  mutate(X1 = gsub("^.*bcl2fastq", "bcl2fastq", X1)) %>%
  mutate(X1 = gsub(" --", "\n\t--", X1)) %>%
  mutate(X1 = gsub(" -", "\n\t-", X1)) %>%
  pull()

```

```{r load read numbers}

trimmin <- read_delim(adaptertrimming_fp) %>%
  filter(!is.na(TrimmedBases))

# read counts by sample number
fastq_sum <- read_delim(fastqsum_fp) %>%
  group_by(SampleNumber) %>%
  #PF means reads that have passed filter
  summarise(total_raw_reads = sum(NumberOfReadsRaw),
            read_counts = sum(NumberOfReadsPF)) %>%
  ungroup()

# SampleSheet.csv loading

sample_sheet <- load_sample_sheet(SampleSheet_fp)

samp_sh_header <- data.frame(X1 = unlist(sample_sheet$Header)) %>%
  mutate(col_names = gsub(",.*", "", X1)) %>%
  mutate(col_names = gsub(" ", "_", col_names)) %>%
  mutate(X1 = gsub(".*,", "", X1)) %>%
  pivot_wider(names_from = "col_names", values_from = "X1")

samp_sh_s <- data.frame(X1 = unlist(sample_sheet$Data), stringsAsFactors = FALSE) %>%
  #split each line into columns using row 1 as the column names
  separate(col = "X1", strsplit(as.character(slice(., 1)), ",")[[1]], sep = ",") %>%
  slice(-1) %>%
  rename(SampleName = "Sample_Name", SampleNumber = "BASESPACE_ONLY_ORIGINAL_SAMPLE_NUMBER") %>%
  mutate(Barcode_sequence = paste0(I7_Index_ID, "+", I5_Index_ID)) %>%
  select(SampleName, SampleNumber, I7_Index_ID, I5_Index_ID, Barcode_sequence, Sample_Type, Index_ID)

#merge sample names with read counts
samp_reads <- read_delim(demuxsum_fp, col_names = FALSE) %>%
  filter(grepl("Sample", X1)) %>%
  t() %>%
  as.data.frame(stringsAsFactors = FALSE) %>%
  purrr::set_names(as.character(slice(., 1))) %>%
  slice(-1) %>%
  merge(fastq_sum, by = "SampleNumber", all = TRUE) %>%
  merge(samp_sh_s, by = c("SampleName", "SampleNumber"), all = TRUE) %>%
  rename(sample_id = "SampleName") %>%
  mutate(percent_reads_passed_filter = read_counts/total_raw_reads) %>%
  mutate(percent_of_lane = read_counts/sum(read_counts)) %>%
  mutate(Sample_Type = ifelse(is.na(Sample_Type), "NA", Sample_Type))

top_unkwn_indices <- read_delim(demuxsum_fp, col_names = FALSE) %>%
  filter(is.na(X3)) %>%
  select(X1, X2) %>%
  filter(!is.na(X2)) %>%
  rename(index_sequence = "X1", number_of_reads = "X2") %>%
  mutate(number_of_reads = as.numeric(number_of_reads)) %>%
  order_on_other_col(index_sequence, number_of_reads) %>%
  slice(1:10) %>%
  droplevels()

#no exhaustive barcode sequence info and no FASTQC info? and no host read mapping info?

```

```{r load data}

kmer <- read_csv(kmer_fp) %>%
  mutate(`SARS-CoV-2` = factor(`SARS-CoV-2`, levels = c("ND", "Detected")))

n_clade <- read_csv(nextclade_fp) %>%
  rename(nc_QC_status = "QC Status")

pango <- read_csv(pangolin_fp) %>%
  rename(pango_QC_status = "QC Status")

bs_results <- merge(kmer, pango, by.x = "Sample", by.y = "Taxon", all = TRUE) %>%
  merge(n_clade, by.x = "Sample", by.y = "seqName", all = TRUE) %>%
  rename(sample_id = "Sample")
#TODO: Add a check to see if positive control shows up as the correct variant


#look for Vibrio campbellii taxa as positive control
#has_pos_ctrl <- any(grepl("Vibrio campbellii", a))

panogolin_database_version <- gsub("^.*v", "", unique(bs_results$Version)[!is.na(unique(bs_results$Version))])
lineage_assignment_software_version <- gsub(paste0("v", panogolin_database_version), "", unique(bs_results$Version)[!is.na(unique(bs_results$Version))])

```

```{r sample_sheet_import and columns, echo=FALSE}
lookup <- c(sample_id = "SampleID",
            subject_id = "SubjectID",
            host_species = "HostSpecies",
            sample_type = "Sample_Type",
            final_library_conc_ng_ul = "final_library_concentration_ng_ul")
ctrl_regex <- "^EBneg.*|^Extract.*|^Vibriolambda.*|^Blank.*|^MockDNA.*|^DNAfreewater.*|^Geneblock.*|^Emptywell|Control"

s_meta <- read_delim(mapping_file_fp, delim = ",") %>%
  #the following lines can be deleted once the code has been finalized
  mutate(plate_row = ifelse(1:n() <= 12, "A", "B")) %>%
  mutate(plate_column = rep(1:12, 2)[1:n()]) %>%
  mutate(plate = 1) %>%
  mutate(plate_coordinate = paste0(plate_row, plate_column)) %>%
  mutate(final_library_conc_ng_ul = rnorm(n(), 20, 10)) %>%
  mutate(dna_extract_conc_ng_ul = rnorm(n(), 10, 3)) %>%
  mutate(ct = rnorm(n(), 20, 15))

s <- s_meta %>%
  merge(samp_reads, by = "sample_id", all = TRUE) %>%
  merge(bs_results, by = "sample_id", all = TRUE) %>%
  rename(any_of(lookup)) %>%
  filter(!grepl("#", sample_id)) %>%
  mutate(isControl = grepl(ctrl_regex, sample_id, ignore.case = TRUE)|grepl(ctrl_regex, sample_type, ignore.case = TRUE)) %>%
  #rearrange the sampletypes by ctrls first
  mutate(sample_type = factor(sample_type, levels =
                               unique(sample_type)[order(match(unique(sample_type), unique(sample_type)[grepl(ctrl_regex, gsub("-| ", "", unique(sample_type)), ignore.case = TRUE)]))])) %>%
  droplevels()

color_by <- NULL
shape_by <- NULL
quality_by <- "sample_type"
potential_headers <- c("study_group", "study_day", "subject_id",
                       "current_antibiotics", "host_species", "cage_number") #pick 2
header_idx <- which(is.element(potential_headers, colnames(s)))

if(length(header_idx)>0){
  color_by <- potential_headers[header_idx[1]]
}
if(length(header_idx)>1){
  shape_by <- potential_headers[header_idx[2]]
}

show.text <- nrow(s) > 40
```


```{r for basic report}
### ====================================
### Select specific groups to plot for QC
### ====================================
s_toPlot <- s %>%
  filter(sample_id != "None") %>%
  mutate(sample_id = as.character(sample_id)) %>%
  droplevels()

```

```{r colors}

ann_colors <- list()

#variables to test and colors
sg_toTest <- c("sample_type")

MononokeStart <- 3

for(g in sg_toTest) {
  lvls <- levels(s_toPlot[, g])
  g_colors <- setNames(ghibli_palette("MononokeLight")[MononokeStart:(MononokeStart-1+length(lvls))], lvls)
  ann_colors[[g]] <- g_colors
}

#scales::show_col(ann_colors[[sg_toTest]])

```

```{r empty plate to plot}
empty_plate <- data.frame(plate_row = unlist(lapply(LETTERS[1:8], function(x) rep(x, 12))), plate_column = rep(1:12, 8), plate = 1) %>%
  mutate(plate_coordinate = paste0(plate_row, plate_column))

```

```{r}
```
