---
title: |
  ![](../aux_files/forKnitting/logo_blk.png){width=4.5in}  
  `r paste0('QC report for sequencing run')`
  `r basename(here())`
author: "Philadelphia Public Health Labs"
date: \today
geometry: margin=3cm
output:
    pdf_document:
        template: ../aux_files/forKnitting/toc_after.tex
        keep_tex: false
        toc: true
        toc_depth: 3
        includes:
            in_header: ../aux_files/forKnitting/TeX_packages_commands.sty

---

\newpage

\tableofcontents

```{r filter out samples, message = TRUE, include=FALSE, eval = FALSE}

#this script filters the Excel file provided by the epidemiologist by excluding low RLU samples and samples received by the Health Centers
#the Excel file should be placed in the extra_metadata folder along with the environmental swabs file
#any samples from previous runs included should be placed in the prev_run folder, using the filtered Excel file
source("code/1_remove_low_RLU.R")

```

```{r generate SampleSheet and metadata, message = TRUE, include=FALSE, eval = FALSE}

#this script generates the SampleSheet.csv file required for demultiplexing the sequencing run
#the required file to include is the sequencing_metadata_sheet.xlsx filled out by the sequencing wet lab scientists, pertaining to the barcodes assigned to the samples
source("code/2_generate_barcodes_IDT.R")

```

```{r upload sequencing run data, message = TRUE, include=FALSE, eval = FALSE}

#this script compresses the sequencing run folder and uploads it to S3, along with auxiliary files

source("code/3_archive_upload_run.R")

```

```{r run nextflow, message = TRUE, include=FALSE, eval = FALSE}

#this script submits the nextflow jobs for demultiplexing and running the cecret pipeline to the ec2 instance

source("code/4_run_nextflow.R")

```

```{r setup for the pdf, echo=FALSE}

library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=FALSE,
  cache.lazy = FALSE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE, #Set message to NA in chunks where message should be sent to stderr
  dpi=100,
  fig.width=6,
  fig.height=6,
  dev.args = list(pdf = list(useDingbats = FALSE))
)

#this lets you loop through variables and produce multiple pander tables and ggplots in a single code block!
#this also means that you have to put results='asis' in any block that has pander output
pander::panderOptions("knitr.auto.asis", FALSE)

```

```{r load data, child = 'code/preamble.Rmd', eval = load_preamble}

```

```{r make figures, child = 'code/data_analysis.Rmd'}

```

```{r generate time stamped report, message=TRUE, include=FALSE, eval=FALSE}
#generate the pdf report

### =============
### Manual inputs
### =============

# need to import the data into a single tsv for the first time?
import_data <- TRUE

# have the original sample accession numbers file to report the results to DDC?
have_phi <- TRUE

### =================
### End manual inputs
### =================

rstudioapi::documentSaveAll()
library(here)
dir.create(here("output"))
project_name <- basename(here())

qc_report_fp <- paste0(project_name, "_QC_Report.Rmd")
qc_output_fp <- here("output", paste0(project_name, '.QC.report_gen.on.', Sys.Date(), '.pdf'))

# load preamble the first time?
load_preamble <- TRUE

# print table as kable for pdf?
use_kable <- TRUE
rmarkdown::render(qc_report_fp, output_file = qc_output_fp)

load_preamble <- FALSE
use_kable <- FALSE
rmarkdown::render(qc_report_fp, output_file = "README.md",
                 output_format = rmarkdown::md_document(variant = "gfm", toc = TRUE))

```

```{r run R script, include=FALSE, eval=FALSE}

rstudioapi::documentSaveAll()
library(here)

#this file needs to sit in a [aux_files/config] directory path above this project directory
tryCatch(
  {
    source(file.path(dirname(here()), "aux_files", "config", "config_variables.R"))
  },
  error = function(e) {
  stop (simpleError("The config_variables.R file needs to sit in a [aux_files/config] directory path above this project directory"))
  }
)

#run this script if exists
tryCatch(
  {
    source(file.path(dirname(here()), "cp2admin.R"))
  }, error = function(e)
  {

  }
)

```

```{r upload data if results look good, message=TRUE, include=FALSE, eval=FALSE}

### =============
### Manual inputs
### =============

#do a test upload first. change this to FALSE when SRA upload and gisaid upload testing is successful and you want to submit the files for real
test_upload <- TRUE

#change this flag to FALSE to not use proxy
use_proxy <- FALSE 

#change this to TRUE to overwrite the submitted files on the sFTP and to resubmit
overwrite <- FALSE

### =================
### End manual inputs
### =================

rstudioapi::documentSaveAll()
library(here)
library(readr)
library(tidyverse)
library(reticulate)

#initialize the seqsender conda environment if it hasn't been already
if(!grepl("seqsender", Sys.which("python"))) {
  use_condaenv("seqsender", required = TRUE)
  import("os")
}

#load variables
tryCatch(
  {
    source(file.path(dirname(here()), "aux_files", "config", "config_variables.R"))
  },
  error = function(e) {
  stop (simpleError("The config_variables.R file needs to sit in a [aux_files/config] directory path above this project directory"))
  }
)
#load functions
tryCatch(
  {
    source(file.path(dirname(here()), "aux_files", "functions", "R_all_functions_v3.R"))
  },
  error = function(e) {
  stop (simpleError("The R_all_functions_v3.R file needs to sit in a [aux_files/functions] directory path above this project directory"))
  }
)

project_name <- basename(here())
seqsender_py <- file.path(seqsender_fp, "seqsender.py")
aux_seqsender_fp <- file.path(dirname(here()), "aux_files", "seqsender")
seqsender_meta_fp <- here("upload", "seqsender", paste(project_name, "PHL2_seqsender_upload.tsv", sep = "_"))
fasta_fp <- here("upload", "fasta", paste0(project_name, "_PHL2_combined.fasta"))

seqsender <- read_tsv(seqsender_meta_fp)

#path of seqsender config file
seqsender_config_fp <- list.files(aux_seqsender_fp, pattern = "seqsender_config.yaml", full.names = TRUE)
seqsender_config_test_fp <- list.files(aux_seqsender_fp, pattern = "seqsender_config_test.yaml", full.names = TRUE)

if(test_upload) {
  test_flag <- "--test"
  config2use <- seqsender_config_test_fp
  gs_client_id <- "TEST-EA76875B00C3"
  gs_log <- here("gisaid", paste0(project_name, "_gisaid_test.log"))
} else{
  test_flag <- ""
  config2use <- seqsender_config_fp
  gs_client_id <- gisaid_client_id
  gs_log <- here("gisaid", paste0(project_name, "_gisaid.log"))
}

if(use_proxy) {
  proxy_flag <- "--proxy"
} else{
  proxy_flag <- ""
}

if(overwrite) {
  overwrite_flag <- "--overwrite"
} else{
  overwrite_flag <- ""
}

file_create_success_msg <- c(paste0("Processing ", project_name, "."),
                             "Processing Files.", 
                             "Creating GISAID files.",
                             "Creating Genbank files.",
                             "Creating BioSample/SRA files.",
                             paste0(project_name, " complete."))

### generate GISAID files to upload. This will also generate test submissions for SRA and BioSample but they will be rewritten with the submit command
upload_files_created <- cli_submit("python",
                                   seqsender_py,
                                   c(proxy_flag, "prep",
                                     "--unique_name", project_name,
                                     "--config", shQuote(config2use, type = "cmd"),
                                     "--metadata", shQuote(seqsender_meta_fp, type = "cmd"),
                                     "--fasta", shQuote(fasta_fp, type = "cmd")),
                                   shQuote_type = "cmd")

if(!all(file_create_success_msg %in% upload_files_created)) {
  stop(simpleError("Seqsender failed to create some of the GISAID upload files!"))
}

# check the seqsender file for GISAID accession numbers
# if the seqsender file already has legit GISAID accession numbers, don't resubmit because GISAID already has these samples
if(all(grepl("^EPI_ISL_0$", seqsender$gisaid_accession)) | all(is.na(seqsender$gisaid_accession))) {
  
  # clear these GISAID log files if exist
  file.remove(gs_log)
  
  # submit files to gisaid
  run_in_terminal(paste("covCLI", "upload",
                        "--clientid", gs_client_id,
                        "--username", gisaid_username,
                        "--metadata", here("gisaid", paste0(project_name, "_gisaid.csv")),
                        "--fasta", here("gisaid", paste0(project_name, "_gisaid.fsa")),
                        "--frameshift", "catch_novel",
                        "--log", gs_log))
  
}
rstudioapi::executeCommand('activateConsole')

# read in the GISAID log file
# if upload was successful, grab the GISAID accessions for the SRA and GenBank submissions
read_gisaid_accessions <- read_csv(gs_log, col_names = "log")

gisaid_errors <- read_gisaid_accessions %>%
  filter(grepl("error", log, ignore.case = TRUE))

if(nrow(gisaid_errors) > 0) {
  stop(simpleError(paste0("Something went wrong! GISAID upload failed. Did you use the wrong credentials?\n",
                          "Otherwise, there may have been a proxy error\n",
                          "This error may also occur if these samples have been already uploaded\n",
                          "Depending on following error message, you may be able to just rerun this chunk to resubmit:\n",
                          gisaid_errors)))
}

gisaid_success <- read_gisaid_accessions %>%
  filter(grepl("^SUCCESS; ", log))

if(nrow(gisaid_success) != nrow(seqsender)) {
  stop(simpleError("Some GISAID samples failed to upload."))
}
  
gisaid_accessions <- read_gisaid_accessions %>%
  filter(grepl("^SUCCESS; ", log)) %>%
  mutate(log = gsub("^SUCCESS; ", "", log)) %>%
  separate(log, into = c("gisaid_name", "gisaid_accession"), sep = " assigned to accession: ", extra = "merge") %>%
  mutate(genbank_note = paste0("GISAID virus name: ", gisaid_name, "; GISAID accession: ", gisaid_accession))
  
if(!test_upload) {
  #if there is a connection error for the real submission, start appending to the file, instead of rewriting it each time. Find a way to keep the column name the same
  gisaid_accessions %>%
    write_csv(here("gisaid", paste0(project_name, "_gisaid_accessions.csv")))
}

#rewrite the seqsender file with the gisaid accession numbers and remove the old prepared files
seqsender %>%
  select(-c(gisaid_accession, genbank_note)) %>%
  merge(gisaid_accessions, by = "gisaid_name", all.x = TRUE) %>%
  write_tsv(seqsender_meta_fp)

### submit files to biosample and sra
cli_submit("python",
           seqsender_py,
           c(proxy_flag, "submit",
             "--unique_name", project_name,
             "--config", shQuote(config2use, type = "cmd"),
             "--metadata", shQuote(seqsender_meta_fp, type = "cmd"),
             "--fasta", shQuote(fasta_fp, type = "cmd"),
             test_flag, overwrite_flag),
           shQuote_type = "cmd")

### submit files to genbank
cli_submit("python",
           seqsender_py,
           c(proxy_flag, "genbank",
             "--unique_name", project_name,
             "--config", shQuote(config2use, type = "cmd"),
             test_flag, overwrite_flag),
           shQuote_type = "cmd")
```
